{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dev work- attempting to work with interpro/pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# system dependecies\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# library dependencies\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "\n",
    "## biopython\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SearchIO\n",
    "\n",
    "\n",
    "# local dependencies/utils\n",
    "\n",
    "## Paths\n",
    "PFAM_PATH = Path(\"/Users/humoodalanzi/pfam/Pfam-A.hmm\")\n",
    "ID_DB_PATH = Path(\"/Users/humoodalanzi/pfam/proteins_id.zip\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Will read API and work on this. HMMER website if interpro is slow.\n",
    "# I need to install GNU make. All of this is done in my test_env which has HMMER and biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_database = pd.read_csv(\"/Users/humoodalanzi/pfam/proteins_id.zip\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6712218, 2)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_database.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2449783"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_database[\"identity\"].isna().sum()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Quick calculations, aside\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.36497369423937065"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2449783/6712218 # the proteins with NaN are 30%!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4262435"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "6712218 - 2449783"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_db = protein_database.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4262435, 2)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_db.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MKFEIKRERLVEGLNDVMKAVSSKTTIPILTGIKMDVSSDGMRLTG...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MLREIGIETEYITLGQLLKMTDTISSGGMAKWFLGEHEVFVNGEAE...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRIDRLELVNYRNYESLELDFSPEINVFIGENAQGKTNIMESLYVL...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTPFDRGRIAALQEEGKTQTAIAEAIGCAKSTISRELKRGTVTQMK...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSQLGTDRVKRGMAEMQKGGVIMDVVNAEQARIAEYAGATAVMALE...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712202</th>\n",
       "      <td>MLKLKVEGMTCGHCAQTVTKAVEAVPAVERALVDLKAGEVTIEGSA...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712203</th>\n",
       "      <td>MNIGQAAKASGVNAKLIRYYESIGLIPEAGRTASGYRVYTPHEVNI...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712204</th>\n",
       "      <td>MRLNSRILNAIAALMTGASLLATSGIALADAGHKHGIGEPAKASAA...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712216</th>\n",
       "      <td>MPCQVYCIGIDASTQFLDTHGLPGHTRRRLPNSPESHGELATILSA...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712217</th>\n",
       "      <td>MSSNDPIQWHGTTILCVRKGGEVVIAGDGQVSMGQTVMKSNARKVR...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4262435 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq identity\n",
       "0        MKFEIKRERLVEGLNDVMKAVSSKTTIPILTGIKMDVSSDGMRLTG...        m\n",
       "1        MLREIGIETEYITLGQLLKMTDTISSGGMAKWFLGEHEVFVNGEAE...        m\n",
       "2        MRIDRLELVNYRNYESLELDFSPEINVFIGENAQGKTNIMESLYVL...        m\n",
       "3        MTPFDRGRIAALQEEGKTQTAIAEAIGCAKSTISRELKRGTVTQMK...        m\n",
       "4        MSQLGTDRVKRGMAEMQKGGVIMDVVNAEQARIAEYAGATAVMALE...        m\n",
       "...                                                    ...      ...\n",
       "6712202  MLKLKVEGMTCGHCAQTVTKAVEAVPAVERALVDLKAGEVTIEGSA...        m\n",
       "6712203  MNIGQAAKASGVNAKLIRYYESIGLIPEAGRTASGYRVYTPHEVNI...        m\n",
       "6712204  MRLNSRILNAIAALMTGASLLATSGIALADAGHKHGIGEPAKASAA...        m\n",
       "6712216  MPCQVYCIGIDASTQFLDTHGLPGHTRRRLPNSPESHGELATILSA...        m\n",
       "6712217  MSSNDPIQWHGTTILCVRKGGEVVIAGDGQVSMGQTVMKSNARKVR...        m\n",
       "\n",
       "[4262435 rows x 2 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "meso_seq_db = protein_db.loc[protein_db[\"identity\"] == \"m\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "thermo_seq_db = protein_db.loc[protein_db[\"identity\"] == \"t\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MKFEIKRERLVEGLNDVMKAVSSKTTIPILTGIKMDVSSDGMRLTG...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MLREIGIETEYITLGQLLKMTDTISSGGMAKWFLGEHEVFVNGEAE...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRIDRLELVNYRNYESLELDFSPEINVFIGENAQGKTNIMESLYVL...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTPFDRGRIAALQEEGKTQTAIAEAIGCAKSTISRELKRGTVTQMK...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSQLGTDRVKRGMAEMQKGGVIMDVVNAEQARIAEYAGATAVMALE...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712202</th>\n",
       "      <td>MLKLKVEGMTCGHCAQTVTKAVEAVPAVERALVDLKAGEVTIEGSA...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712203</th>\n",
       "      <td>MNIGQAAKASGVNAKLIRYYESIGLIPEAGRTASGYRVYTPHEVNI...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712204</th>\n",
       "      <td>MRLNSRILNAIAALMTGASLLATSGIALADAGHKHGIGEPAKASAA...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712216</th>\n",
       "      <td>MPCQVYCIGIDASTQFLDTHGLPGHTRRRLPNSPESHGELATILSA...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6712217</th>\n",
       "      <td>MSSNDPIQWHGTTILCVRKGGEVVIAGDGQVSMGQTVMKSNARKVR...</td>\n",
       "      <td>m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4108740 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq identity\n",
       "0        MKFEIKRERLVEGLNDVMKAVSSKTTIPILTGIKMDVSSDGMRLTG...        m\n",
       "1        MLREIGIETEYITLGQLLKMTDTISSGGMAKWFLGEHEVFVNGEAE...        m\n",
       "2        MRIDRLELVNYRNYESLELDFSPEINVFIGENAQGKTNIMESLYVL...        m\n",
       "3        MTPFDRGRIAALQEEGKTQTAIAEAIGCAKSTISRELKRGTVTQMK...        m\n",
       "4        MSQLGTDRVKRGMAEMQKGGVIMDVVNAEQARIAEYAGATAVMALE...        m\n",
       "...                                                    ...      ...\n",
       "6712202  MLKLKVEGMTCGHCAQTVTKAVEAVPAVERALVDLKAGEVTIEGSA...        m\n",
       "6712203  MNIGQAAKASGVNAKLIRYYESIGLIPEAGRTASGYRVYTPHEVNI...        m\n",
       "6712204  MRLNSRILNAIAALMTGASLLATSGIALADAGHKHGIGEPAKASAA...        m\n",
       "6712216  MPCQVYCIGIDASTQFLDTHGLPGHTRRRLPNSPESHGELATILSA...        m\n",
       "6712217  MSSNDPIQWHGTTILCVRKGGEVVIAGDGQVSMGQTVMKSNARKVR...        m\n",
       "\n",
       "[4108740 rows x 2 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meso_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>122152</th>\n",
       "      <td>MEPVRDGFFDRPTLELAQALLGMLLVCVTEEGAAAGRIVETEAYIG...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122153</th>\n",
       "      <td>MKAVTYQGAHSIEVKEVADPRIERPDDIIVRLTASGICGSDLHLVH...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122154</th>\n",
       "      <td>MANSIQQFKADFFKALAHPLRIRILEILCEGDRSVNEIQNLLRSEG...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122155</th>\n",
       "      <td>MANDELILTPEGYKQLEEELRELKGPKRKALAERLKVAISYGDLRE...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122156</th>\n",
       "      <td>MDWLDRMNRALDYIERHLEDGVSIEEAARAAYSSAFHFHRLFHMVA...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683703</th>\n",
       "      <td>MSGLRQNRSLWETVIPRGLRMSFWSDERRHSQSRSIGRRPIDLETG...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683704</th>\n",
       "      <td>MRADARRYKSDAFLSEAMKMLAVGMIPLSAGACSFSFGTVGLQEEE...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683705</th>\n",
       "      <td>MRIIVHHATCQGHARCWAQAPDIFKLDDEGYILPGDINVAEGDELL...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683706</th>\n",
       "      <td>MNQARISELRDREAIRDCLYRYCRGIDRADEAALRSSYWPDAHDHH...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683708</th>\n",
       "      <td>MRSATTSFWHSDAVRGPSTPTEKGGLIRRLKAGWHGGDRLLCLDGT...</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153695 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               protein_seq identity\n",
       "122152   MEPVRDGFFDRPTLELAQALLGMLLVCVTEEGAAAGRIVETEAYIG...        t\n",
       "122153   MKAVTYQGAHSIEVKEVADPRIERPDDIIVRLTASGICGSDLHLVH...        t\n",
       "122154   MANSIQQFKADFFKALAHPLRIRILEILCEGDRSVNEIQNLLRSEG...        t\n",
       "122155   MANDELILTPEGYKQLEEELRELKGPKRKALAERLKVAISYGDLRE...        t\n",
       "122156   MDWLDRMNRALDYIERHLEDGVSIEEAARAAYSSAFHFHRLFHMVA...        t\n",
       "...                                                    ...      ...\n",
       "6683703  MSGLRQNRSLWETVIPRGLRMSFWSDERRHSQSRSIGRRPIDLETG...        t\n",
       "6683704  MRADARRYKSDAFLSEAMKMLAVGMIPLSAGACSFSFGTVGLQEEE...        t\n",
       "6683705  MRIIVHHATCQGHARCWAQAPDIFKLDDEGYILPGDINVAEGDELL...        t\n",
       "6683706  MNQARISELRDREAIRDCLYRYCRGIDRADEAALRSSYWPDAHDHH...        t\n",
       "6683708  MRSATTSFWHSDAVRGPSTPTEKGGLIRRLKAGWHGGDRLLCLDGT...        t\n",
       "\n",
       "[153695 rows x 2 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "thermo_seq_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26.733075246429618"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "4108740/153695 # ~27-fold more meso than thermo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv(\"../data/Sample.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Old code archive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Define the amino acid sequence\n",
    "# meso_seq_list = [df['meso_seq'].iloc[i] for i in range(1001)] # loop-over 1000 meso sequences\n",
    "\n",
    "# # Create a list of SeqRecord objects\n",
    "# records = []\n",
    "# for i, seq in enumerate(meso_seq_list):\n",
    "#     record = SeqRecord(Seq(seq), id=f\"meso_seq{i+1}\")\n",
    "#     records.append(record)\n",
    "\n",
    "# # Write the list of SeqRecord objects to a FASTA file\n",
    "# SeqIO.write(records, \"sequences.fasta\", \"fasta\")\n",
    "\n",
    "# # Write sequence to file in FASTA format\n",
    "# with open(\"input.fasta\", \"w\") as input_file:\n",
    "#     SeqIO.write(records, input_file, \"fasta\")\n",
    "\n",
    "# # Run HMMER search against the Pfam database using the hmmscan command\n",
    "# subprocess.run([\"hmmscan\", \"--cpu\", \"4\", \"--domtblout\", \"output.domtblout\", \"/Users/humoodalanzi/pfam/Pfam-A.hmm\", \"input.fasta\"])\n",
    "\n",
    "\n",
    "# # Parse the HMMER results file\n",
    "# best_hit = None\n",
    "# best_evalue = float(\"inf\")\n",
    "# best_bitscore = float(\"-inf\")\n",
    "# with open(\"output.domtblout\", \"r\") as results_file:\n",
    "#     for line in results_file:\n",
    "#         if not line.startswith(\"#\"):\n",
    "#             fields = line.strip().split()\n",
    "#             family_id = fields[0]\n",
    "#             evalue = float(fields[6])\n",
    "#             bitscore = float(fields[7])\n",
    "#             devalue = float(fields[11])\n",
    "#             start = int(fields[19])\n",
    "#             end = int(fields[20])\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"family_id\": family_id, \"evalue\": evalue,\"devalue\":devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "\n",
    "\n",
    "# # Print the best hit and its attributes\n",
    "# if best_hit is not None:\n",
    "#     print(\"Best Pfam family match: \" + best_hit[\"family_id\"])\n",
    "#     print(\"Total E-value: \" + str(best_hit[\"evalue\"]))\n",
    "#     print(\"Domain E-vlaue: \" + str(best_hit[\"devalue\"]))\n",
    "#     print(\"Bit score: \" + str(best_hit[\"bitscore\"]))\n",
    "#     print(\"Start position: \" + str(best_hit[\"start\"]))\n",
    "#     print(\"End position: \" + str(best_hit[\"end\"]))\n",
    "# else:\n",
    "#     print(\"No Pfam family match found.\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Get the best hit (hsps: high-scoring segment pair)\n",
    "# best_hit = None\n",
    "# for qresult in qresults:\n",
    "#     for hit in qresult.hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hsp.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.bitscore / hit.seq_len\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "\n",
    "\n",
    "# # Print the best hit and its attributes\n",
    "# if best_hit is not None:\n",
    "#     print(\"Best Pfam family match: \" + best_hit[\"family_id\"])\n",
    "#     print(\"Total E-value: \" + str(best_hit[\"evalue\"]))\n",
    "#     print(\"length of hit sequence or HMM: \" + str(best_hit[\"seqlen\"]))\n",
    "#     print(\"High-scoring pair bias: \" + str(best_hit[\"bias\"]))\n",
    "#     print(\"Domain E-vlaue: \" + str(best_hit[\"devalue\"]))\n",
    "#     print(\"Bit score: \" + str(best_hit[\"bitscore\"]))\n",
    "#     print(\"Start position: \" + str(best_hit[\"start\"]))\n",
    "#     print(\"End position: \" + str(best_hit[\"end\"]))\n",
    "# else:\n",
    "#     print(\"No Pfam family match found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meso_seq_list = df['meso_seq'].iloc[:1000]\n",
    "# meso_seq_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from collections import Counter, defaultdict\n",
    "# import subprocess\n",
    "# from typing import Dict, List, Optional\n",
    "# from Bio import SeqIO\n",
    "# from Bio.SeqRecord import SeqRecord\n",
    "# from Bio.Seq import Seq\n",
    "# from Bio.Alphabet import generic_protein\n",
    "# from Bio import SearchIO\n",
    "\n",
    "# def read_sequences(filename: str, n: Optional[int] = None) -> List[SeqRecord]:\n",
    "#     \"\"\"Reads n protein sequences from a file and returns them as a list of SeqRecord objects.\"\"\"\n",
    "#     records = []\n",
    "#     with open(filename, \"r\") as f:\n",
    "#         for i, record in enumerate(SeqIO.parse(f, \"fasta\")):\n",
    "#             if n is not None and i >= n:\n",
    "#                 break\n",
    "#             records.append(record)\n",
    "#     return records\n",
    "\n",
    "# def write_sequences(records: List[SeqRecord], filename: str) -> None:\n",
    "#     \"\"\"Writes a list of SeqRecord objects to a FASTA file.\"\"\"\n",
    "#     with open(filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "# def run_hmmer(pfam_path: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "#     \"\"\"Runs HMMER search against the Pfam database using the hmmscan command.\"\"\"\n",
    "#     with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         subprocess.run([\"hmmscan\", \"--cpu\", str(num_cpu), \"--domtblout\", output_filename, pfam_path, input_filename], stdout=f)\n",
    "\n",
    "# def find_best_hit(query_id: str, hits: List) -> Optional[Dict]:\n",
    "#     \"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "\n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str\n",
    "#         The ID of the query sequence.\n",
    "#     hits : List\n",
    "#         List of HMMER hits to search through.\n",
    "\n",
    "#     Returns:\n",
    "#     ------------\n",
    "#     dict: A dictionary containing information about the best hit, including family ID, e-value, \n",
    "#           sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "#           If no hit is found, returns None.\n",
    "#     \"\"\"\n",
    "#     best_hit = None\n",
    "#     for hit in hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hit.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.evalue\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "#     return best_hit\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     # Define input and output filenames\n",
    "#     input_filename = \"input.fasta\"\n",
    "#     output_filename = \"output.domtblout\"\n",
    "\n",
    "#     # Define the Pfam database path\n",
    "#     pfam_path = \"/path/to/Pfam-A.hmm\"\n",
    "\n",
    "#     # Read in protein sequences from a FASTA file\n",
    "#     sequences = read_sequences(input_filename)\n",
    "\n",
    "#     # Write protein sequences to a FASTA file\n",
    "#     write_sequences(sequences, input_filename)\n",
    "\n",
    "#     # Run HMMER search against the Pfam database\n",
    "#     run_hmmer(pfam_path, input_filename, output_filename)\n",
    "\n",
    "#     # Parse HMMER results\n",
    "#     hits_dict = defaultdict(list)\n",
    "#     for query in SearchIO.parse(output_filename, 'hmmer3-domtbl'):\n",
    "#         query_id = query.id\n",
    "#         hits = query.hsps\n",
    "#         best_hit = find_best_hit(query_id, hits)\n",
    "#         if best_hit:\n",
    "#             hits_dict[best_hit['family_id']].append(best_hit)\n",
    "\n",
    "#     # Print summary of hits by family\n",
    "#     for family_id, hits in hits_dict.items():\n",
    "#         print(f\"Family ID: {family_id}\")\n",
    "#         print(f\"Number of hits: {len(hits)}\")\n",
    "#         evalues = [hit['evalue'] for hit in hits]\n",
    "#         avg_evalue = sum(evalues) / len(evalues)\n",
    "#         print(f\"Average e-value: {avg_evalue:.2g}\")\n",
    "#         print(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from collections import Counter\n",
    "# import subprocess\n",
    "# from typing import Dict, List, Optional\n",
    "# from Bio import SeqIO\n",
    "# from Bio.SeqRecord import SeqRecord\n",
    "# from Bio.Seq import Seq\n",
    "# from Bio.Alphabet import generic_protein\n",
    "# from Bio import SearchIO\n",
    "\n",
    "\n",
    "# def read_sequences(filename: str, n: Optional[int] = None) -> List[SeqRecord]:\n",
    "#     \"\"\"Reads n protein sequences from a file and returns them as a list of SeqRecord objects.\"\"\"\n",
    "#     records = []\n",
    "#     with open(filename, \"r\") as f:\n",
    "#         for i, record in enumerate(SeqIO.parse(f, \"fasta\")):\n",
    "#             if n is not None and i >= n:\n",
    "#                 break\n",
    "#             records.append(record)\n",
    "#     return records\n",
    "\n",
    "\n",
    "# def write_sequences(records: List[SeqRecord], filename: str) -> None:\n",
    "#     \"\"\"Writes a list of SeqRecord objects to a FASTA file.\"\"\"\n",
    "#     with open(filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "\n",
    "# def run_hmmer(pfam_path: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "#     \"\"\"Runs HMMER search against the Pfam database using the hmmscan command.\"\"\"\n",
    "#     with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
    "#         subprocess.run(\n",
    "#             [\n",
    "#                 \"hmmscan\",\n",
    "#                 \"--cpu\",\n",
    "#                 str(num_cpu),\n",
    "#                 \"--domtblout\",\n",
    "#                 output_filename,\n",
    "#                 pfam_path,\n",
    "#                 input_filename,\n",
    "#             ],\n",
    "#             stdout=f,\n",
    "#         )\n",
    "\n",
    "\n",
    "# def find_best_hit(query_id: str, hits: List) -> Optional[Dict]:\n",
    "#     \"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "\n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str\n",
    "#         The ID of the query sequence.\n",
    "#     hits : List[SearchIO.Hit]\n",
    "#         List of HMMER hits.\n",
    "#     \"\"\"\n",
    "#     hits_by_evalue = sorted(hits, key=lambda hit: hit.evalue)\n",
    "#     for hit in hits_by_evalue:\n",
    "#         if hit.id == query_id:\n",
    "#             continue\n",
    "#         return {\n",
    "#             \"query_id\": query_id,\n",
    "#             \"hit_id\": hit.id,\n",
    "#             \"evalue\": hit.evalue,\n",
    "#             \"bitscore\": hit.bitscore,\n",
    "#         }\n",
    "#     return None\n",
    "\n",
    "\n",
    "# def count_domains(hits: List[SearchIO.Hit]) -> Dict[str, int]:\n",
    "#     \"\"\"Counts the frequency of each Pfam domain in a list of HMMER hits.\"\"\"\n",
    "#     domains = Counter(hit.accession.split(\".\")[0] for hit in hits)\n",
    "#     return domains\n",
    "\n",
    "\n",
    "# def main():\n",
    "#     # Paths to input/output files and the Pfam database\n",
    "#     input_filename = \"input.fasta\"\n",
    "#     output_filename = \"output.tbl\"\n",
    "#     pfam_path = \"/path/to/Pfam-A.hmm\"\n",
    "\n",
    "#     # Read input sequences from file\n",
    "#     input_seqs = read_sequences(input_filename)\n",
    "\n",
    "#     # Write input sequences to file\n",
    "#     input_seqs_file = \"input_seqs.fasta\"\n",
    "#     write_sequences(input_seqs, input_seqs_file)\n",
    "\n",
    "#     # Run HMMER search against the Pfam database\n",
    "#     run_hmmer(pfam_path, input_seqs_file, output_filename)\n",
    "\n",
    "#     # Parse the HMMER output\n",
    "#     hits = []\n",
    "#     with open(output_filename, \"r\") as f:\n",
    "#         for qresult in SearchIO.parse(f, \"hmmer3-tab\"):\n",
    "#             for hit in qresult.hits:\n",
    "#                 hits.append(hit)\n",
    "\n",
    "#     # Count the frequency of each Pfam domain in the hits\n",
    "#     domain_counts = count_domains(hits)\n",
    "\n",
    "#     # Find the best hit for each input sequence\n",
    "#     best_hits\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimizing the code"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Things I want to do:\n",
    "- Neeed to make the list of AA sequence a function, i.e. more modualr \n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<_io.TextIOWrapper name='output_test.domtblout' mode='w' encoding='utf-8'>"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run_hmmer(PFAM_PATH, \"test.fasta\", \"output_test.domtblout\", 4) # works!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query sequence ID: 0\n",
      "Best Pfam family match: DNA_pol3_beta\n",
      "Total E-value: 1.6e-45\n",
      "Length of hit sequence or HMM: 121\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.3e-31\n",
      "Bit score: 109.3\n",
      "Start position: 0\n",
      "End position: 127\n",
      "----\n",
      "Query sequence ID: 1\n",
      "Best Pfam family match: S4_2\n",
      "Total E-value: 4.4e-27\n",
      "Length of hit sequence or HMM: 65\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 5e-27\n",
      "Bit score: 93.7\n",
      "Start position: 5\n",
      "End position: 69\n",
      "----\n",
      "Query sequence ID: 2\n",
      "Best Pfam family match: SMC_N\n",
      "Total E-value: 1e-18\n",
      "Length of hit sequence or HMM: 220\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 1.9e-18\n",
      "Bit score: 66.8\n",
      "Start position: 2\n",
      "End position: 350\n",
      "----\n",
      "Query sequence ID: 3\n",
      "Best Pfam family match: HTH_38\n",
      "Total E-value: 4.2e-12\n",
      "Length of hit sequence or HMM: 44\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.1e-11\n",
      "Bit score: 44.4\n",
      "Start position: 0\n",
      "End position: 40\n",
      "----\n",
      "Query sequence ID: 4\n",
      "Best Pfam family match: SOR_SNZ\n",
      "Total E-value: 5.7e-104\n",
      "Length of hit sequence or HMM: 207\n",
      "High-scoring pair bias: 5.9\n",
      "Domain E-vlaue: 7.5e-104\n",
      "Bit score: 346.1\n",
      "Start position: 6\n",
      "End position: 212\n",
      "----\n",
      "Query sequence ID: 5\n",
      "Best Pfam family match: SNO\n",
      "Total E-value: 2.6e-71\n",
      "Length of hit sequence or HMM: 188\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2.9e-71\n",
      "Bit score: 239.3\n",
      "Start position: 5\n",
      "End position: 187\n",
      "----\n",
      "Query sequence ID: 6\n",
      "Best Pfam family match: Abhydrolase_1\n",
      "Total E-value: 2.1e-25\n",
      "Length of hit sequence or HMM: 257\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.3e-24\n",
      "Bit score: 87.4\n",
      "Start position: 26\n",
      "End position: 271\n",
      "----\n",
      "Query sequence ID: 7\n",
      "Best Pfam family match: MafB19-deam\n",
      "Total E-value: 1.1e-35\n",
      "Length of hit sequence or HMM: 138\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.3e-35\n",
      "Bit score: 122.2\n",
      "Start position: 8\n",
      "End position: 152\n",
      "----\n",
      "Query sequence ID: 8\n",
      "Best Pfam family match: YitT_membrane\n",
      "Total E-value: 1.7e-64\n",
      "Length of hit sequence or HMM: 206\n",
      "High-scoring pair bias: 19.6\n",
      "Domain E-vlaue: 2.2e-64\n",
      "Bit score: 216.8\n",
      "Start position: 15\n",
      "End position: 218\n",
      "----\n",
      "Query sequence ID: 9\n",
      "Best Pfam family match: dNK\n",
      "Total E-value: 2.6e-49\n",
      "Length of hit sequence or HMM: 204\n",
      "High-scoring pair bias: 0.9\n",
      "Domain E-vlaue: 2.9e-49\n",
      "Bit score: 167.8\n",
      "Start position: 5\n",
      "End position: 203\n",
      "----\n",
      "Query sequence ID: 10\n",
      "Best Pfam family match: dNK\n",
      "Total E-value: 5.8e-64\n",
      "Length of hit sequence or HMM: 204\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 6.6e-64\n",
      "Bit score: 215.7\n",
      "Start position: 14\n",
      "End position: 212\n",
      "----\n",
      "Query sequence ID: 11\n",
      "Best Pfam family match: YbaB_DNA_bd\n",
      "Total E-value: 7.1e-33\n",
      "Length of hit sequence or HMM: 90\n",
      "High-scoring pair bias: 3.2\n",
      "Domain E-vlaue: 8.5e-33\n",
      "Bit score: 112.6\n",
      "Start position: 10\n",
      "End position: 100\n",
      "----\n",
      "Query sequence ID: 12\n",
      "Best Pfam family match: Toprim_4\n",
      "Total E-value: 2.3e-25\n",
      "Length of hit sequence or HMM: 89\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 3.6e-25\n",
      "Bit score: 88.3\n",
      "Start position: 80\n",
      "End position: 171\n",
      "----\n",
      "Query sequence ID: 13\n",
      "Best Pfam family match: Thymidylate_kin\n",
      "Total E-value: 5.3e-50\n",
      "Length of hit sequence or HMM: 185\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 6.1e-50\n",
      "Bit score: 169.6\n",
      "Start position: 6\n",
      "End position: 195\n",
      "----\n",
      "Query sequence ID: 14\n",
      "Best Pfam family match: CdAMP_rec\n",
      "Total E-value: 2.1e-53\n",
      "Length of hit sequence or HMM: 109\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2.3e-53\n",
      "Bit score: 179.1\n",
      "Start position: 0\n",
      "End position: 109\n",
      "----\n",
      "Query sequence ID: 15\n",
      "Best Pfam family match: DNA_pol3_delta2\n",
      "Total E-value: 6.1e-58\n",
      "Length of hit sequence or HMM: 161\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.1e-57\n",
      "Bit score: 194.8\n",
      "Start position: 11\n",
      "End position: 170\n",
      "----\n",
      "Query sequence ID: 16\n",
      "Best Pfam family match: PSP1\n",
      "Total E-value: 6.8e-31\n",
      "Length of hit sequence or HMM: 86\n",
      "High-scoring pair bias: 2.9\n",
      "Domain E-vlaue: 1.1e-30\n",
      "Bit score: 105.9\n",
      "Start position: 60\n",
      "End position: 146\n",
      "----\n",
      "Query sequence ID: 17\n",
      "Best Pfam family match: YabA\n",
      "Total E-value: 5.6e-31\n",
      "Length of hit sequence or HMM: 102\n",
      "High-scoring pair bias: 0.7\n",
      "Domain E-vlaue: 6.6e-31\n",
      "Bit score: 107.1\n",
      "Start position: 1\n",
      "End position: 103\n",
      "----\n",
      "Query sequence ID: 18\n",
      "Best Pfam family match: MTS\n",
      "Total E-value: 5.9e-11\n",
      "Length of hit sequence or HMM: 170\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 4.8e-10\n",
      "Bit score: 39.4\n",
      "Start position: 15\n",
      "End position: 145\n",
      "----\n",
      "Query sequence ID: 19\n",
      "Best Pfam family match: GIY-YIG\n",
      "Total E-value: 2.5e-17\n",
      "Length of hit sequence or HMM: 78\n",
      "High-scoring pair bias: 0.6\n",
      "Domain E-vlaue: 3.1e-17\n",
      "Bit score: 62.7\n",
      "Start position: 5\n",
      "End position: 77\n",
      "----\n"
     ]
    }
   ],
   "source": [
    "# parse_hmmer(\"output_test.domtblout\") # works!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Scrap\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# records = []\n",
    "# for i, seq in enumerate(meso_seq_list):\n",
    "#     record = SeqRecord(Seq(seq), id=str(i))\n",
    "#     records.append(record)\n",
    "\n",
    "# # Write the list of SeqRecord objects to a FASTA file\n",
    "# with open(\"input.fasta\", \"w\") as f:\n",
    "#     SeqIO.write(records, f, \"fasta\")\n",
    "\n",
    "\n",
    "# # Run HMMER search against the Pfam database using the hmmscan command\n",
    "# with open(\"output.domtblout\", \"w\") as f:\n",
    "#     subprocess.run([\"hmmscan\", \"--cpu\", \"4\", \"--domtblout\", \"output.domtblout\", PFAM_PATH, \"input.fasta\"], stdout=f)\n",
    "\n",
    "# # Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "# def find_best_hit(query_id, hits):\n",
    "#     \"\"\"\"\n",
    "#     Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "#     Parameters:\n",
    "#     ------------\n",
    "#     query_id : str \n",
    "#         The ID of the query sequence.\n",
    "#     hits : list \n",
    "#         List of HMMER hits to search through.\n",
    "    \n",
    "#     Returns:\n",
    "#     ------------\n",
    "#     dict :\n",
    "#         A dictionary containing information about the best hit, including family ID, e-value, \n",
    "#         sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "#         If no hit is found, returns None.\n",
    "    \n",
    "#     Raises\n",
    "#     -------\n",
    "    \n",
    "#     \"\"\"\n",
    "#     best_hit = None\n",
    "#     for hit in hits:\n",
    "#         for hsp in hit.hsps:\n",
    "#             family_id = hit.id\n",
    "#             evalue = hit.evalue\n",
    "#             seqlen = hit.seq_len\n",
    "#             bitscore = hsp.bitscore\n",
    "#             devalue = hsp.evalue\n",
    "#             bias = hsp.bias\n",
    "#             start = hsp.query_start\n",
    "#             end = hsp.query_end\n",
    "#             if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "#                 best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "#     return best_hit\n",
    "\n",
    "# # Parse the HMMER results file using HmmerIO\n",
    "# with open(\"output.domtblout\", \"r\") as results_file:\n",
    "#     qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "# # Store the counts of hits for each family ID\n",
    "# hit_counts = defaultdict(int)\n",
    "# for qresult in qresults:\n",
    "#     query_id = qresult.id\n",
    "#     best_hit = find_best_hit(query_id, qresult.hits)\n",
    "#     if best_hit is not None:\n",
    "#         family_id = best_hit[\"family_id\"]\n",
    "#         hit_counts[family_id] += 1\n",
    "#         print(f\"Query sequence ID: {query_id}\")\n",
    "#         print(f\"Best Pfam family match: {family_id}\")\n",
    "#         print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "#         print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "#         print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "#         print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "#         print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "#         print(f\"Start position: {best_hit['start']}\")\n",
    "#         print(f\"End position: {best_hit['end']}\")\n",
    "#         print(\"----\")\n",
    "#     else:\n",
    "#         print(\"No Pfam family match found for query sequence ID:\", query_id)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Final version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a list of SeqRecord objects\n",
    "# Write the list of SeqRecord objects to a FASTA file\n",
    "def read_seq(lists: pd.core.series.Series, inputname: str = \"input\"):\n",
    "    \"\"\"\n",
    "    Returns a list of SeqRecord objects and creates a corresponding input Fasta of them\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "    list : pandas.core.series.Series\n",
    "        a list of string amino acid sequences\n",
    "    input name : str, default = 'input'\n",
    "        a name for the input fasta file\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    fasta : \n",
    "        an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    records = []\n",
    "    for i, seq in enumerate(lists):\n",
    "        record = SeqRecord(Seq(seq), id=str(i))\n",
    "        records.append(record)\n",
    "    \n",
    "    with open(f\"{inputname}.fasta\", \"w\") as file:\n",
    "            SeqIO.write(records, file, \"fasta\")\n",
    "    return file\n",
    "\n",
    "def run_hmmer(DB_PATH: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "    \"\"\"\n",
    "    Runs HMMER search against the Pfam database using the hmmscan command\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    DB_PATH : str\n",
    "        the path of the HMMER/pfam database\n",
    "    input_filename : str\n",
    "        a name for the input fasta file (don't forget extension!)\n",
    "    output_filename : str\n",
    "        a name for the domtblout output file (don't forget extension!)\n",
    "    num_cpu : int\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    domtblout : \n",
    "        an output domtblout file of the HMMER/pfam results\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    with open(output_filename, \"w\", encoding=\"utf-8\") as file:\n",
    "        subprocess.run([\"hmmscan\", \"--cpu\", str(num_cpu), \"--domtblout\", output_filename, DB_PATH, input_filename], stdout=file)\n",
    "    return file\n",
    "\n",
    "# Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "def find_best_hit(query_id, hits):\n",
    "    \"\"\"\"\n",
    "    Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "    Parameters:\n",
    "    ------------\n",
    "    query_id : str \n",
    "        The ID of the query sequence.\n",
    "    hits : list \n",
    "        List of HMMER hits to search through.\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    dict :\n",
    "        A dictionary containing information about the best hit, including family ID, e-value, \n",
    "        sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "        If no hit is found, returns None.\n",
    "    \n",
    "    Raises\n",
    "    -------\n",
    "    \n",
    "    \"\"\"\n",
    "    best_hit = None\n",
    "    for hit in hits:\n",
    "        for hsp in hit.hsps:\n",
    "            family_id = hit.id\n",
    "            evalue = hit.evalue\n",
    "            seqlen = hit.seq_len\n",
    "            bitscore = hsp.bitscore\n",
    "            devalue = hsp.evalue\n",
    "            bias = hsp.bias\n",
    "            start = hsp.query_start\n",
    "            end = hsp.query_end\n",
    "            if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "                best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "    return best_hit\n",
    "\n",
    "# Parse the HMMER results file using HmmerIO\n",
    "# Store the counts of hits for each family ID\n",
    "def parse_hmmer(output_name: str):\n",
    "    \"\"\"\n",
    "    parses HMMER output result\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "    output_name : str\n",
    "        a name for the domtblout output file (don't forget extension!)\n",
    "\n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "    file : \n",
    "        an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    # Parse the HMMER results file using HmmerIO\n",
    "    with open(output_name, \"r\") as results_file:\n",
    "        qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "    # Store the counts of hits for each family ID\n",
    "    hit_counts = defaultdict(int)\n",
    "    for qresult in qresults:\n",
    "        query_id = qresult.id\n",
    "        best_hit = find_best_hit(query_id, qresult.hits)\n",
    "        if best_hit is not None:\n",
    "            family_id = best_hit[\"family_id\"]\n",
    "            hit_counts[family_id] += 1\n",
    "            print(f\"Query sequence ID: {query_id}\")\n",
    "            print(f\"Best Pfam family match: {family_id}\")\n",
    "            print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "            print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "            print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "            print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "            print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "            print(f\"Start position: {best_hit['start']}\")\n",
    "            print(f\"End position: {best_hit['end']}\")\n",
    "            print(\"----\")\n",
    "        else:\n",
    "            print(\"No Pfam family match found for query sequence ID:\", query_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Execution (it works)\n",
    "\n",
    "# Define the amino acid sequence\n",
    "meso_seq_list = meso_seq_db[\"protein_seq\"].iloc[:15]\n",
    "thermo_seq_list = thermo_seq_db[\"protein_seq\"].iloc[:15]\n",
    "\n",
    "\n",
    "# generate meso and thermo files\n",
    "read_seq(meso_seq_list, \"meso_input\")\n",
    "read_seq(thermo_seq_list, \"thermo_input\")\n",
    "\n",
    "# place files into HMMER/pfam\n",
    "run_hmmer(PFAM_PATH, \"meso_input.fasta\", \"meso_output.domtblout\", 4)\n",
    "run_hmmer(PFAM_PATH, \"thermo_input.fasta\", \"thermo_output.domtblout\", 4)\n",
    "\n",
    "# parse outputs of HMMER/pfam\n",
    "parse_hmmer(\"meso_output.domtblout\")\n",
    "parse_hmmer(\"thermo_output.domtblout\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Attempting wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hmmer_wrapper(seq: pd.core.series.Series, input_filename: str, pfam_path: str, input_filename_with_ext: str, output_filename_with_ext: str, cpu: int = 4):\n",
    "    \"\"\"\n",
    "    A wrapper function that executes the HMMER against pfam and parsing the results\n",
    "\n",
    "    Parameters:\n",
    "    ------------\n",
    "        \n",
    "    \n",
    "    Returns:\n",
    "    ------------\n",
    "\n",
    "    Raises\n",
    "    -------\n",
    "    TypeError\n",
    "    \"\"\"\n",
    "    # Create a list of SeqRecord objects\n",
    "    # Write the list of SeqRecord objects to a FASTA file\n",
    "    def read_seq(lists: pd.core.series.Series, inputname: str = \"input\"):\n",
    "        \"\"\"\n",
    "        Returns a list of SeqRecord objects and creates a corresponding input Fasta of them\n",
    "\n",
    "        Parameters:\n",
    "        ------------\n",
    "        list : pandas.core.series.Series\n",
    "            a list of string amino acid sequences\n",
    "        input name : str, default = 'input'\n",
    "            a name for the input fasta file\n",
    "\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        fasta : \n",
    "            an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "        Raises\n",
    "        -------\n",
    "        TypeError\n",
    "        \"\"\"\n",
    "        records = []\n",
    "        for i, seq in enumerate(lists):\n",
    "            record = SeqRecord(Seq(seq), id=str(i))\n",
    "            records.append(record)\n",
    "    \n",
    "        with open(f\"{inputname}.fasta\", \"w\") as file:\n",
    "            SeqIO.write(records, file, \"fasta\")\n",
    "        return file\n",
    "\n",
    "    def run_hmmer(DB_PATH: str, input_filename: str, output_filename: str, num_cpu: int = 4) -> None:\n",
    "        \"\"\"\n",
    "        Runs HMMER search against the Pfam database using the hmmscan command\n",
    "    \n",
    "        Parameters:\n",
    "        ------------\n",
    "        DB_PATH : str\n",
    "            the path of the HMMER/pfam database\n",
    "        input_filename : str\n",
    "            a name for the input fasta file (don't forget extension!)\n",
    "        output_filename : str\n",
    "            a name for the domtblout output file (don't forget extension!)\n",
    "        num_cpu : int\n",
    "\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        domtblout : \n",
    "            an output domtblout file of the HMMER/pfam results\n",
    "\n",
    "        Raises\n",
    "        -------\n",
    "        TypeError\n",
    "        \"\"\"\n",
    "        with open(output_filename, \"w\", encoding=\"utf-8\") as file:\n",
    "            subprocess.run([\"hmmscan\", \"--cpu\", str(num_cpu), \"--domtblout\", output_filename, DB_PATH, input_filename], stdout=file)\n",
    "        return file\n",
    "\n",
    "    # Define a function to find the best hit (hsps: high-scoring segment pair)\n",
    "    def find_best_hit(query_id, hits):\n",
    "        \"\"\"\"\n",
    "        Given a list of HMMER hits, returns the best hit based on e-value.\n",
    "    \n",
    "        Parameters:\n",
    "        ------------\n",
    "        query_id : str \n",
    "            The ID of the query sequence.\n",
    "        hits : list \n",
    "            List of HMMER hits to search through.\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        dict :\n",
    "            A dictionary containing information about the best hit, including family ID, e-value, \n",
    "            sequence length, bit score, high-scoring pair bias, domain e-value, start position, and end position.\n",
    "            If no hit is found, returns None.\n",
    "    \n",
    "        Raises\n",
    "        -------\n",
    "    \n",
    "        \"\"\"\n",
    "        best_hit = None\n",
    "        for hit in hits:\n",
    "            for hsp in hit.hsps:\n",
    "                family_id = hit.id\n",
    "                evalue = hit.evalue\n",
    "                seqlen = hit.seq_len\n",
    "                bitscore = hsp.bitscore\n",
    "                devalue = hsp.evalue\n",
    "                bias = hsp.bias\n",
    "                start = hsp.query_start\n",
    "                end = hsp.query_end\n",
    "                if best_hit is None or evalue < best_hit[\"evalue\"]:\n",
    "                    best_hit = {\"query_id\": query_id, \"family_id\": family_id, \"evalue\": evalue,\"seqlen\": seqlen ,\"bias\": bias, \"devalue\": devalue, \"bitscore\": bitscore, \"start\": start, \"end\": end}\n",
    "        return best_hit\n",
    "\n",
    "    # Parse the HMMER results file using HmmerIO\n",
    "    # Store the counts of hits for each family ID\n",
    "    def parse_hmmer(output_name: str):\n",
    "        \"\"\"\n",
    "        parses HMMER output result\n",
    "\n",
    "        Parameters:\n",
    "        ------------\n",
    "        output_name : str\n",
    "            a name for the domtblout output file (don't forget extension!)\n",
    "\n",
    "    \n",
    "        Returns:\n",
    "        ------------\n",
    "        file : \n",
    "            an input fasta file from a list of SeqRecord objects of originally string the amino acid sequences\n",
    "\n",
    "        Raises\n",
    "        -------\n",
    "        TypeError\n",
    "        \"\"\"\n",
    "        # Parse the HMMER results file using HmmerIO\n",
    "        with open(output_name, \"r\") as results_file:\n",
    "            qresults = list(SearchIO.parse(results_file, 'hmmscan3-domtab'))\n",
    "\n",
    "        # Store the counts of hits for each family ID\n",
    "        hit_counts = defaultdict(int)\n",
    "        for qresult in qresults:\n",
    "            query_id = qresult.id\n",
    "            best_hit = find_best_hit(query_id, qresult.hits)\n",
    "            if best_hit is not None:\n",
    "                family_id = best_hit[\"family_id\"]\n",
    "                hit_counts[family_id] += 1\n",
    "                print(f\"Query sequence ID: {query_id}\")\n",
    "                print(f\"Best Pfam family match: {family_id}\")\n",
    "                print(f\"Total E-value: {best_hit['evalue']}\")\n",
    "                print(f\"Length of hit sequence or HMM: {best_hit['seqlen']}\")\n",
    "                print(f\"High-scoring pair bias: {best_hit['bias']}\")\n",
    "                print(f\"Domain E-vlaue: {best_hit['devalue']}\")\n",
    "                print(f\"Bit score: {best_hit['bitscore']}\")\n",
    "                print(f\"Start position: {best_hit['start']}\")\n",
    "                print(f\"End position: {best_hit['end']}\")\n",
    "                print(\"----\")\n",
    "            else:\n",
    "                print(\"No Pfam family match found for query sequence ID:\", query_id)\n",
    "    \n",
    "    ## execution\n",
    "    # User should define an amino acid sequence as Pandas series\n",
    "    \n",
    "    # generate meso and thermo files\n",
    "    read_seq(seq, input_filename)\n",
    "\n",
    "    # place files into HMMER/pfam\n",
    "    run_hmmer(pfam_path, input_filename_with_ext, output_filename_with_ext, cpu)\n",
    "\n",
    "    # parse outputs of HMMER/pfam\n",
    "    parse_hmmer(output_filename_with_ext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query sequence ID: 0\n",
      "Best Pfam family match: DNA_pol3_beta\n",
      "Total E-value: 1.6e-45\n",
      "Length of hit sequence or HMM: 121\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.3e-31\n",
      "Bit score: 109.3\n",
      "Start position: 0\n",
      "End position: 127\n",
      "----\n",
      "Query sequence ID: 1\n",
      "Best Pfam family match: S4_2\n",
      "Total E-value: 4.4e-27\n",
      "Length of hit sequence or HMM: 65\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 5e-27\n",
      "Bit score: 93.7\n",
      "Start position: 5\n",
      "End position: 69\n",
      "----\n",
      "Query sequence ID: 2\n",
      "Best Pfam family match: SMC_N\n",
      "Total E-value: 1e-18\n",
      "Length of hit sequence or HMM: 220\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 1.9e-18\n",
      "Bit score: 66.8\n",
      "Start position: 2\n",
      "End position: 350\n",
      "----\n",
      "Query sequence ID: 3\n",
      "Best Pfam family match: HTH_38\n",
      "Total E-value: 4.2e-12\n",
      "Length of hit sequence or HMM: 44\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.1e-11\n",
      "Bit score: 44.4\n",
      "Start position: 0\n",
      "End position: 40\n",
      "----\n",
      "Query sequence ID: 4\n",
      "Best Pfam family match: SOR_SNZ\n",
      "Total E-value: 5.7e-104\n",
      "Length of hit sequence or HMM: 207\n",
      "High-scoring pair bias: 5.9\n",
      "Domain E-vlaue: 7.5e-104\n",
      "Bit score: 346.1\n",
      "Start position: 6\n",
      "End position: 212\n",
      "----\n",
      "Query sequence ID: 5\n",
      "Best Pfam family match: SNO\n",
      "Total E-value: 2.6e-71\n",
      "Length of hit sequence or HMM: 188\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2.9e-71\n",
      "Bit score: 239.3\n",
      "Start position: 5\n",
      "End position: 187\n",
      "----\n",
      "Query sequence ID: 6\n",
      "Best Pfam family match: Abhydrolase_1\n",
      "Total E-value: 2.1e-25\n",
      "Length of hit sequence or HMM: 257\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.3e-24\n",
      "Bit score: 87.4\n",
      "Start position: 26\n",
      "End position: 271\n",
      "----\n",
      "Query sequence ID: 7\n",
      "Best Pfam family match: MafB19-deam\n",
      "Total E-value: 1.1e-35\n",
      "Length of hit sequence or HMM: 138\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.3e-35\n",
      "Bit score: 122.2\n",
      "Start position: 8\n",
      "End position: 152\n",
      "----\n",
      "Query sequence ID: 8\n",
      "Best Pfam family match: YitT_membrane\n",
      "Total E-value: 1.7e-64\n",
      "Length of hit sequence or HMM: 206\n",
      "High-scoring pair bias: 19.6\n",
      "Domain E-vlaue: 2.2e-64\n",
      "Bit score: 216.8\n",
      "Start position: 15\n",
      "End position: 218\n",
      "----\n",
      "Query sequence ID: 9\n",
      "Best Pfam family match: dNK\n",
      "Total E-value: 2.6e-49\n",
      "Length of hit sequence or HMM: 204\n",
      "High-scoring pair bias: 0.9\n",
      "Domain E-vlaue: 2.9e-49\n",
      "Bit score: 167.8\n",
      "Start position: 5\n",
      "End position: 203\n",
      "----\n",
      "Query sequence ID: 10\n",
      "Best Pfam family match: dNK\n",
      "Total E-value: 5.8e-64\n",
      "Length of hit sequence or HMM: 204\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 6.6e-64\n",
      "Bit score: 215.7\n",
      "Start position: 14\n",
      "End position: 212\n",
      "----\n",
      "Query sequence ID: 11\n",
      "Best Pfam family match: YbaB_DNA_bd\n",
      "Total E-value: 7.1e-33\n",
      "Length of hit sequence or HMM: 90\n",
      "High-scoring pair bias: 3.2\n",
      "Domain E-vlaue: 8.5e-33\n",
      "Bit score: 112.6\n",
      "Start position: 10\n",
      "End position: 100\n",
      "----\n",
      "Query sequence ID: 12\n",
      "Best Pfam family match: Toprim_4\n",
      "Total E-value: 2.3e-25\n",
      "Length of hit sequence or HMM: 89\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 3.6e-25\n",
      "Bit score: 88.3\n",
      "Start position: 80\n",
      "End position: 171\n",
      "----\n",
      "Query sequence ID: 13\n",
      "Best Pfam family match: Thymidylate_kin\n",
      "Total E-value: 5.3e-50\n",
      "Length of hit sequence or HMM: 185\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 6.1e-50\n",
      "Bit score: 169.6\n",
      "Start position: 6\n",
      "End position: 195\n",
      "----\n",
      "Query sequence ID: 14\n",
      "Best Pfam family match: CdAMP_rec\n",
      "Total E-value: 2.1e-53\n",
      "Length of hit sequence or HMM: 109\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2.3e-53\n",
      "Bit score: 179.1\n",
      "Start position: 0\n",
      "End position: 109\n",
      "----\n",
      "Query sequence ID: 15\n",
      "Best Pfam family match: DNA_pol3_delta2\n",
      "Total E-value: 6.1e-58\n",
      "Length of hit sequence or HMM: 161\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.1e-57\n",
      "Bit score: 194.8\n",
      "Start position: 11\n",
      "End position: 170\n",
      "----\n",
      "Query sequence ID: 16\n",
      "Best Pfam family match: PSP1\n",
      "Total E-value: 6.8e-31\n",
      "Length of hit sequence or HMM: 86\n",
      "High-scoring pair bias: 2.9\n",
      "Domain E-vlaue: 1.1e-30\n",
      "Bit score: 105.9\n",
      "Start position: 60\n",
      "End position: 146\n",
      "----\n",
      "Query sequence ID: 17\n",
      "Best Pfam family match: YabA\n",
      "Total E-value: 5.6e-31\n",
      "Length of hit sequence or HMM: 102\n",
      "High-scoring pair bias: 0.7\n",
      "Domain E-vlaue: 6.6e-31\n",
      "Bit score: 107.1\n",
      "Start position: 1\n",
      "End position: 103\n",
      "----\n",
      "Query sequence ID: 18\n",
      "Best Pfam family match: MTS\n",
      "Total E-value: 5.9e-11\n",
      "Length of hit sequence or HMM: 170\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 4.8e-10\n",
      "Bit score: 39.4\n",
      "Start position: 15\n",
      "End position: 145\n",
      "----\n",
      "Query sequence ID: 19\n",
      "Best Pfam family match: GIY-YIG\n",
      "Total E-value: 2.5e-17\n",
      "Length of hit sequence or HMM: 78\n",
      "High-scoring pair bias: 0.6\n",
      "Domain E-vlaue: 3.1e-17\n",
      "Bit score: 62.7\n",
      "Start position: 5\n",
      "End position: 77\n",
      "----\n",
      "Query sequence ID: 20\n",
      "Best Pfam family match: TP_methylase\n",
      "Total E-value: 1.9e-33\n",
      "Length of hit sequence or HMM: 213\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2.7e-33\n",
      "Bit score: 115.8\n",
      "Start position: 12\n",
      "End position: 213\n",
      "----\n",
      "Query sequence ID: 21\n",
      "Best Pfam family match: AbrB_C\n",
      "Total E-value: 5.5e-16\n",
      "Length of hit sequence or HMM: 37\n",
      "High-scoring pair bias: 0.3\n",
      "Domain E-vlaue: 1e-15\n",
      "Bit score: 57.2\n",
      "Start position: 52\n",
      "End position: 89\n",
      "----\n",
      "Query sequence ID: 22\n",
      "Best Pfam family match: TatD_DNase\n",
      "Total E-value: 6.1e-89\n",
      "Length of hit sequence or HMM: 255\n",
      "High-scoring pair bias: 0.3\n",
      "Domain E-vlaue: 7e-89\n",
      "Bit score: 297.7\n",
      "Start position: 2\n",
      "End position: 253\n",
      "----\n",
      "Query sequence ID: 23\n",
      "Best Pfam family match: DUF4093\n",
      "Total E-value: 1.1e-27\n",
      "Length of hit sequence or HMM: 83\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 2.1e-27\n",
      "Bit score: 95.5\n",
      "Start position: 92\n",
      "End position: 177\n",
      "----\n",
      "Query sequence ID: 24\n",
      "Best Pfam family match: RrnaAD\n",
      "Total E-value: 1.7e-66\n",
      "Length of hit sequence or HMM: 265\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 2e-66\n",
      "Bit score: 223.9\n",
      "Start position: 20\n",
      "End position: 278\n",
      "----\n",
      "Query sequence ID: 25\n",
      "Best Pfam family match: VEG\n",
      "Total E-value: 6.2e-30\n",
      "Length of hit sequence or HMM: 73\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 7e-30\n",
      "Bit score: 103.0\n",
      "Start position: 2\n",
      "End position: 75\n",
      "----\n",
      "Query sequence ID: 26\n",
      "Best Pfam family match: GHMP_kinases_N\n",
      "Total E-value: 3.7e-12\n",
      "Length of hit sequence or HMM: 66\n",
      "High-scoring pair bias: 0.6\n",
      "Domain E-vlaue: 7.6e-12\n",
      "Bit score: 45.4\n",
      "Start position: 84\n",
      "End position: 142\n",
      "----\n",
      "Query sequence ID: 27\n",
      "Best Pfam family match: PuR_N\n",
      "Total E-value: 8.1e-32\n",
      "Length of hit sequence or HMM: 70\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 2.4e-31\n",
      "Bit score: 107.6\n",
      "Start position: 3\n",
      "End position: 73\n",
      "----\n",
      "Query sequence ID: 28\n",
      "Best Pfam family match: Ribonuc_L-PSP\n",
      "Total E-value: 1.9e-43\n",
      "Length of hit sequence or HMM: 121\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 2.2e-43\n",
      "Bit score: 147.2\n",
      "Start position: 7\n",
      "End position: 122\n",
      "----\n",
      "Query sequence ID: 29\n",
      "Best Pfam family match: SpoVG\n",
      "Total E-value: 1.2e-36\n",
      "Length of hit sequence or HMM: 82\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 1.4e-36\n",
      "Bit score: 124.7\n",
      "Start position: 0\n",
      "End position: 82\n",
      "----\n",
      "Query sequence ID: 30\n",
      "Best Pfam family match: Pribosyltran_N\n",
      "Total E-value: 1.5e-53\n",
      "Length of hit sequence or HMM: 116\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 1.7e-52\n",
      "Bit score: 176.4\n",
      "Start position: 9\n",
      "End position: 126\n",
      "----\n",
      "Query sequence ID: 31\n",
      "Best Pfam family match: Ribosomal_TL5_C\n",
      "Total E-value: 2e-21\n",
      "Length of hit sequence or HMM: 85\n",
      "High-scoring pair bias: 0.4\n",
      "Domain E-vlaue: 3.5e-21\n",
      "Bit score: 75.5\n",
      "Start position: 100\n",
      "End position: 183\n",
      "----\n",
      "Query sequence ID: 32\n",
      "Best Pfam family match: Pept_tRNA_hydro\n",
      "Total E-value: 8.2e-69\n",
      "Length of hit sequence or HMM: 183\n",
      "High-scoring pair bias: 0.0\n",
      "Domain E-vlaue: 9.1e-69\n",
      "Bit score: 231.0\n",
      "Start position: 2\n",
      "End position: 185\n",
      "----\n",
      "Query sequence ID: 33\n",
      "Best Pfam family match: S4\n",
      "Total E-value: 1.6e-11\n",
      "Length of hit sequence or HMM: 48\n",
      "High-scoring pair bias: 0.4\n",
      "Domain E-vlaue: 2.8e-11\n",
      "Bit score: 43.1\n",
      "Start position: 0\n",
      "End position: 47\n",
      "----\n",
      "Query sequence ID: 34\n",
      "Best Pfam family match: OTCace\n",
      "Total E-value: 3.6e-58\n",
      "Length of hit sequence or HMM: 157\n",
      "High-scoring pair bias: 0.1\n",
      "Domain E-vlaue: 5.4e-58\n",
      "Bit score: 195.6\n",
      "Start position: 157\n",
      "End position: 312\n",
      "----\n"
     ]
    }
   ],
   "source": [
    "##### Testing wrapper\n",
    "meso_seq_list = meso_seq_db[\"protein_seq\"].iloc[:35]\n",
    "thermo_seq_list = thermo_seq_db[\"protein_seq\"].iloc[:35]\n",
    "\n",
    "hmmer_wrapper(meso_seq_list, \"meso_input\", PFAM_PATH, \"meso_input.fasta\", \"meso_output.domtblout\", 4)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Put best pfam metrics as a column in the original DB make sure they correspond to the original IDs"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ideas for unit tests:\n",
    "\n",
    "- empty pfam input\n",
    "- collections key error\n",
    "```py \n",
    "from collections import defaultdict\n",
    "# create a defaultdict that returns 0 as default value\n",
    "my_dict = defaultdict(int)\n",
    "\n",
    "# add some values\n",
    "my_dict['a'] = 1\n",
    "my_dict['b'] = 2\n",
    "\n",
    "# try to access a nonexistent key\n",
    "print(my_dict['c'])  # prints 0, instead of raising KeyError\n",
    "```\n",
    "- pfam output error handling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "4a9b5cea0784dd84a54eb32a9034ea0f7d299d9a522482e4c348f8e08377d626"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
