{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "265e8129",
   "metadata": {},
   "source": [
    "This notebook uses iFeatureOmega, a feature generation software, to add to our feature space for a RandomForestClassifier that predicts protein pair functionality."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b88f847",
   "metadata": {},
   "source": [
    "To do:\n",
    "\n",
    "1) Write unit tests for iFeatureOmega\n",
    "2) time trial for different descriptors - DONE\n",
    "3) Figure out how to append meso and thermo descriptors\n",
    "4) Make sure protein length is in training\n",
    "5) try changing subject_align_len to subject_align_len/m_protein_len\n",
    "6) log(ratio) for proteins approacing zero or infinity\n",
    "    take distributions to assess for this.\n",
    "    Update -- taking log and then droping infinites and NaN improves score to 0.86 <-- cannot repeat this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "50c23d28",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'path_to_db' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-44-2e633829120f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mduckdb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mcon\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mduckdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath_to_db\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\"SELECT list_of_column_names FROM fafsa_final\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcon\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'path_to_db' is not defined"
     ]
    }
   ],
   "source": [
    "import duckdb\n",
    "\n",
    "con = duckdb.connect(path_to_db)\n",
    "cmd = \"\"\"SELECT list_of_column_names FROM fafsa_final\"\"\"\n",
    "df = con.execute(cmd).df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa9845b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df.to_parquet(path)\n",
    "cmd = f\"\"\"CREATE OR REPLACE TABLE table_name AS SELECT * FROM '{path}'\"\"\"\n",
    "con.execute(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19973c41",
   "metadata": {},
   "source": [
    " protein_pair_cmd = \"\"\"CREATE OR REPLACE TABLE fafsa_protein_pairs AS\n",
    "                          SELECT \n",
    "                          meso_pid,\n",
    "                          thermo_pid,\n",
    "                          bit_score,\n",
    "                          local_gap_compressed_percent_id,\n",
    "                          scaled_local_query_percent_id,\n",
    "                          scaled_local_symmetric_percent_id,\n",
    "                          query_align_len,\n",
    "                          query_align_cov,\n",
    "                          subject_align_len,\n",
    "                          subject_align_cov,\n",
    "                          FROM protein_pairs USING SAMPLE 100\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "56e8db3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import sklearn.preprocessing\n",
    "import sklearn.model_selection\n",
    "import sklearn.ensemble\n",
    "import sklearn.feature_selection\n",
    "import unittest\n",
    "import iFeatureOmegaCLI\n",
    "import Bio.SeqIO\n",
    "import Bio.SeqRecord\n",
    "import io\n",
    "from io import StringIO\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "99990cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fasta_from_dataframe(dataframe, output_file_a, output_file_b):\n",
    "    #adjust this to write function with BioPython\n",
    "    #separate functions for each of the input sequences\n",
    "    #in training, seq_a = meso and seq_b = thermo\n",
    "    \n",
    "    \n",
    "    #meso sequence to fasta\n",
    "    with open(output_file_a, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), row['m_protein_seq']))\n",
    "    \n",
    "    #thermo sequence to fasta\n",
    "    with open(output_file_b, 'w') as f:\n",
    "        for _, row in df.iterrows():\n",
    "            f.write('>{}\\n{}\\n'.format((row['prot_pair_index']), (row['t_protein_seq'])))\n",
    "   \n",
    "    #return output files\n",
    "    return [output_file_a, output_file_b]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "653eb96d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_descriptors(fasta_file, descriptors=[]):\n",
    "    \n",
    "    \"\"\"\n",
    "    Generates features from a protein sequence\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Fasta file with amino acid sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Vector of descriptors\n",
    "    \"\"\"\n",
    "    \n",
    "    #create iProtein object\n",
    "    protein = iFeatureOmegaCLI.iProtein(fasta_file)\n",
    "    \n",
    "    #not sure why we need this yet. Right now it is stored in local directory.\n",
    "    params = protein.import_parameters('protein_parameters.json')\n",
    "    \n",
    "    protein_descriptors = {}\n",
    "    \n",
    "    for descriptor in descriptors:\n",
    "        protein.get_descriptor(descriptor)\n",
    "        protein_descriptors.update({f'{descriptor}':protein.encodings})\n",
    "        \n",
    "    return protein_descriptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "fe48f193",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_new_dataframe(dataframe, output_files, descriptors=[]):\n",
    "    \"\"\"\n",
    "    Creates new dataframe with descriptors added.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    Pandas dataframe, list of descriptors as strings, output file name.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Dataframe including vector(s) of descriptors\n",
    "    \"\"\"\n",
    "\n",
    "    fasta_files = get_fasta_from_dataframe(dataframe, output_files[0], output_files[1])\n",
    "    \n",
    "    def compute_descriptor_ratio(fasta_files, descriptors=[]):\n",
    "        \"\"\"\n",
    "        Generates dictionary of descriptors for each of the two input sequences.\n",
    "        Computes the difference between each instance of a descriptor.\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        List of two fasta files (str) and list of descriptors (str).\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        Dictionary with difference between descriptors for each of the \n",
    "        input sequences.\n",
    "        \"\"\"\n",
    "        desc_a = get_protein_descriptors(fasta_files[0], descriptors)\n",
    "        desc_b = get_protein_descriptors(fasta_files[1], descriptors)\n",
    "\n",
    "        feature_dict = {}\n",
    "\n",
    "        for key in desc_a:\n",
    "            \n",
    "            if 'AAC' in key:\n",
    "                feature_dict[key] = desc_a[key] - desc_b[key]\n",
    "            elif 'GAAC' in key:\n",
    "                feature_dict[key] = desc_a[key] - desc_b[key]\n",
    "            else:\n",
    "                feature_dict[key] = desc_a[key] / desc_b[key]\n",
    "\n",
    "        return feature_dict\n",
    "    \n",
    "    feature_dict = compute_descriptor_ratio(fasta_files, descriptors)\n",
    "\n",
    "\n",
    "    df = dataframe.reset_index()\n",
    "\n",
    "    for desc in descriptors:\n",
    "\n",
    "        feature_dict[desc].index = feature_dict[desc].index.astype(int)\n",
    "        features = feature_dict[desc].reset_index()\n",
    "\n",
    "        df = pd.merge(\n",
    "            df,\n",
    "            features,\n",
    "            how='outer',\n",
    "            left_index=True,\n",
    "            right_index=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "38c1120f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#iFeature properties\n",
    "\n",
    "# protein = iFeatureOmegaCLI.iProtein('meso_50k.fasta')\n",
    "# protein.display_feature_types()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ba87d47f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/FAFSA\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/FAFSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "f90c1cdd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'local_gap_compressed_percent_id',\n",
       "       'scaled_local_query_percent_id', 'scaled_local_symmetric_percent_id',\n",
       "       'query_align_len', 'query_align_cov', 'subject_align_len',\n",
       "       'subject_align_cov', 'bit_score', 'thermo_index', 'meso_index',\n",
       "       'prot_pair_index', 'meso_protein_int_index', 'thermo_protein_int_index',\n",
       "       'taxa_pair_index', 'local_gap_compressed_percent_id_16s',\n",
       "       'scaled_local_query_percent_id_16s',\n",
       "       'scaled_local_symmetric_percent_id_16s', 'query_align_cov_16s',\n",
       "       'subject_align_cov_16s', 'bit_score_16s', 'm_ogt', 't_ogt',\n",
       "       'ogt_difference', 'm_protein_seq', 't_protein_seq', 'm_protein_desc',\n",
       "       't_protein_desc', 'm_protein_len', 't_protein_len'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#convert to pandas df\n",
    "df = pd.read_csv('learn2therm_sample_50k.csv')\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "d7f0a563",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/loganroberts/Learn2Therm/ValidProt/notebooks\n"
     ]
    }
   ],
   "source": [
    "cd /Users/loganroberts/Learn2Therm/ValidProt/notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "043bf4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "target = pd.read_csv('protein_match_6k.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "4dfd5bc5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>thermo_index</th>\n",
       "      <th>...</th>\n",
       "      <th>bit_score_16s</th>\n",
       "      <th>m_ogt</th>\n",
       "      <th>t_ogt</th>\n",
       "      <th>ogt_difference</th>\n",
       "      <th>m_protein_seq</th>\n",
       "      <th>t_protein_seq</th>\n",
       "      <th>m_protein_desc</th>\n",
       "      <th>t_protein_desc</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.287582</td>\n",
       "      <td>0.217822</td>\n",
       "      <td>0.215686</td>\n",
       "      <td>160</td>\n",
       "      <td>0.792079</td>\n",
       "      <td>152</td>\n",
       "      <td>0.737864</td>\n",
       "      <td>131</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1153.0</td>\n",
       "      <td>27.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.5</td>\n",
       "      <td>MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...</td>\n",
       "      <td>MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...</td>\n",
       "      <td>ECF RNA polymerase sigma factor SigK</td>\n",
       "      <td>sigma-70 family RNA polymerase sigma factor</td>\n",
       "      <td>206</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.319635</td>\n",
       "      <td>0.295359</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>218</td>\n",
       "      <td>0.919831</td>\n",
       "      <td>226</td>\n",
       "      <td>0.969957</td>\n",
       "      <td>282</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>1014.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...</td>\n",
       "      <td>MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>233</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.279621</td>\n",
       "      <td>0.234127</td>\n",
       "      <td>0.218924</td>\n",
       "      <td>211</td>\n",
       "      <td>0.837302</td>\n",
       "      <td>210</td>\n",
       "      <td>0.731707</td>\n",
       "      <td>96</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...</td>\n",
       "      <td>MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>SDR family oxidoreductase</td>\n",
       "      <td>287</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>875</td>\n",
       "      <td>...</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...</td>\n",
       "      <td>MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...</td>\n",
       "      <td>ectoine/hydroxyectoine ABC transporter permeas...</td>\n",
       "      <td>amino acid ABC transporter permease</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.338710</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.287671</td>\n",
       "      <td>60</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>71</td>\n",
       "      <td>0.887500</td>\n",
       "      <td>61</td>\n",
       "      <td>9827</td>\n",
       "      <td>...</td>\n",
       "      <td>991.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...</td>\n",
       "      <td>MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>YqzK family protein</td>\n",
       "      <td>80</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49995</th>\n",
       "      <td>49995</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.169972</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>180</td>\n",
       "      <td>0.509915</td>\n",
       "      <td>200</td>\n",
       "      <td>0.738007</td>\n",
       "      <td>53</td>\n",
       "      <td>14963</td>\n",
       "      <td>...</td>\n",
       "      <td>1055.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...</td>\n",
       "      <td>MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...</td>\n",
       "      <td>aminoglycoside phosphotransferase family protein</td>\n",
       "      <td>serine/threonine protein kinase</td>\n",
       "      <td>271</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49996</th>\n",
       "      <td>49996</td>\n",
       "      <td>0.417989</td>\n",
       "      <td>0.389163</td>\n",
       "      <td>0.392060</td>\n",
       "      <td>190</td>\n",
       "      <td>0.935961</td>\n",
       "      <td>187</td>\n",
       "      <td>0.935000</td>\n",
       "      <td>314</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>1083.0</td>\n",
       "      <td>26.5</td>\n",
       "      <td>52.5</td>\n",
       "      <td>26.0</td>\n",
       "      <td>MFRTGVKAEIGRSLAVVGEAEDVERAVRVVLEQRPDVVLLDVHLPG...</td>\n",
       "      <td>MILEAEPDIVVVGEAGDGEKAVEEARALQPDVVLMDIRMPRKDGVE...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>200</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49997</th>\n",
       "      <td>49997</td>\n",
       "      <td>0.352174</td>\n",
       "      <td>0.239645</td>\n",
       "      <td>0.242152</td>\n",
       "      <td>230</td>\n",
       "      <td>0.680473</td>\n",
       "      <td>227</td>\n",
       "      <td>0.685801</td>\n",
       "      <td>279</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>991.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>24.5</td>\n",
       "      <td>MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...</td>\n",
       "      <td>MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>331</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49998</th>\n",
       "      <td>49998</td>\n",
       "      <td>0.334764</td>\n",
       "      <td>0.331915</td>\n",
       "      <td>0.329810</td>\n",
       "      <td>232</td>\n",
       "      <td>0.987234</td>\n",
       "      <td>230</td>\n",
       "      <td>0.966387</td>\n",
       "      <td>281</td>\n",
       "      <td>11324</td>\n",
       "      <td>...</td>\n",
       "      <td>907.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>MSESHAGALLSVRGLTAGYGGATALDGVSLTVAAGETVALLGANGA...</td>\n",
       "      <td>MSLLTTSGLTRHFSGIHAVEGVDFTLEAGEIRALIGSNGAGKTTLV...</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>ABC transporter ATP-binding protein</td>\n",
       "      <td>238</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49999</th>\n",
       "      <td>49999</td>\n",
       "      <td>0.336207</td>\n",
       "      <td>0.184834</td>\n",
       "      <td>0.174497</td>\n",
       "      <td>113</td>\n",
       "      <td>0.535545</td>\n",
       "      <td>117</td>\n",
       "      <td>0.495763</td>\n",
       "      <td>165</td>\n",
       "      <td>7134</td>\n",
       "      <td>...</td>\n",
       "      <td>978.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>26.5</td>\n",
       "      <td>MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...</td>\n",
       "      <td>MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...</td>\n",
       "      <td>response regulator transcription factor</td>\n",
       "      <td>response regulator</td>\n",
       "      <td>236</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50000 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  local_gap_compressed_percent_id  \\\n",
       "0               0                         0.287582   \n",
       "1               1                         0.319635   \n",
       "2               2                         0.279621   \n",
       "3               3                         0.327273   \n",
       "4               4                         0.338710   \n",
       "...           ...                              ...   \n",
       "49995       49995                         0.344828   \n",
       "49996       49996                         0.417989   \n",
       "49997       49997                         0.352174   \n",
       "49998       49998                         0.334764   \n",
       "49999       49999                         0.336207   \n",
       "\n",
       "       scaled_local_query_percent_id  scaled_local_symmetric_percent_id  \\\n",
       "0                           0.217822                           0.215686   \n",
       "1                           0.295359                           0.297872   \n",
       "2                           0.234127                           0.218924   \n",
       "3                           0.200743                           0.214712   \n",
       "4                           0.318182                           0.287671   \n",
       "...                              ...                                ...   \n",
       "49995                       0.169972                           0.192308   \n",
       "49996                       0.389163                           0.392060   \n",
       "49997                       0.239645                           0.242152   \n",
       "49998                       0.331915                           0.329810   \n",
       "49999                       0.184834                           0.174497   \n",
       "\n",
       "       query_align_len  query_align_cov  subject_align_len  subject_align_cov  \\\n",
       "0                  160         0.792079                152           0.737864   \n",
       "1                  218         0.919831                226           0.969957   \n",
       "2                  211         0.837302                210           0.731707   \n",
       "3                  166         0.617100                163           0.696581   \n",
       "4                   60         0.909091                 71           0.887500   \n",
       "...                ...              ...                ...                ...   \n",
       "49995              180         0.509915                200           0.738007   \n",
       "49996              190         0.935961                187           0.935000   \n",
       "49997              230         0.680473                227           0.685801   \n",
       "49998              232         0.987234                230           0.966387   \n",
       "49999              113         0.535545                117           0.495763   \n",
       "\n",
       "       bit_score  thermo_index  ...  bit_score_16s  m_ogt  t_ogt  \\\n",
       "0            131           875  ...         1153.0   27.5   50.0   \n",
       "1            282         11324  ...         1014.0   25.0   54.0   \n",
       "2             96           875  ...         1138.0   28.0   50.0   \n",
       "3            175           875  ...         1077.0   28.0   50.0   \n",
       "4             61          9827  ...          991.0   30.0   50.0   \n",
       "...          ...           ...  ...            ...    ...    ...   \n",
       "49995         53         14963  ...         1055.0   28.0   52.5   \n",
       "49996        314          7134  ...         1083.0   26.5   52.5   \n",
       "49997        279          7134  ...          991.0   28.0   52.5   \n",
       "49998        281         11324  ...          907.0   29.0   54.0   \n",
       "49999        165          7134  ...          978.0   26.0   52.5   \n",
       "\n",
       "       ogt_difference                                      m_protein_seq  \\\n",
       "0                22.5  MAESGTSRRADHLVPVPGPDAEPPAVADELLRAVGRGDEQAFGRLY...   \n",
       "1                29.0  MARIALVDDDRNILTSVSMTLEAEGFEVETYNDGQSALDAFNKRMP...   \n",
       "2                22.0  MKDTVVFVTGAARGIGAHTARLAVARGARVALVGLEPHLLADLAAE...   \n",
       "3                22.0  MTSGLWERVLDGVWVTIQLLVLSALLATAVSFVVGIARTHRLWIVR...   \n",
       "4                20.0  MIISLRRGLRFIRFIVFFAALVYLFYHVLDLFNGWISPVDQYQMPT...   \n",
       "...               ...                                                ...   \n",
       "49995            24.5  MDNATFRLGDDLSVRLPGHSRWIGQVEREQRWLPWLAPRLPLTVST...   \n",
       "49996            26.0  MFRTGVKAEIGRSLAVVGEAEDVERAVRVVLEQRPDVVLLDVHLPG...   \n",
       "49997            24.5  MIRLAELTKTYPGQQHPAVDGISMEVAEGEIVVLVGPSGCGKTTTL...   \n",
       "49998            25.0  MSESHAGALLSVRGLTAGYGGATALDGVSLTVAAGETVALLGANGA...   \n",
       "49999            26.5  MTGQGASGSETGVSDDSPITVVVVDDQELLRTGLRDLAEHDGDIAV...   \n",
       "\n",
       "                                           t_protein_seq  \\\n",
       "0      MPSQITESERIELAERFERDALPLLDQLYSAALRMTRNPADAEDLV...   \n",
       "1      MRVLLVEDDPNTSRSIEMMLTHANLNVYATDMGEEGIDLAKLYDYD...   \n",
       "2      MTPEQIFSGQTAIVTGGASGIGAATVEHIARRGGRVFSVDLSYDSP...   \n",
       "3      MAMSRRKRGQLARGIQYAILVIVVVVLALLADWGKIGKAFFDWEAA...   \n",
       "4      MKRMVWRTLKVFIIFIACTLLFYFGLRFMHLEYEQFHRYEPPEGPA...   \n",
       "...                                                  ...   \n",
       "49995  MPPQPRPLRPNDPREIGGFALLGRLGEGGQGTVYLGGAPDGRRVAV...   \n",
       "49996  MILEAEPDIVVVGEAGDGEKAVEEARALQPDVVLMDIRMPRKDGVE...   \n",
       "49997  MTEQPILSARGLTVDFRLRGGRRARAVDGVDLDLAPGEVLALAGES...   \n",
       "49998  MSLLTTSGLTRHFSGIHAVEGVDFTLEAGEIRALIGSNGAGKTTLV...   \n",
       "49999  MTTEAQPRRVVLAEDEALIRLDLAEMLAEEGYEVVGQAADGETALA...   \n",
       "\n",
       "                                          m_protein_desc  \\\n",
       "0                   ECF RNA polymerase sigma factor SigK   \n",
       "1                response regulator transcription factor   \n",
       "2                              SDR family oxidoreductase   \n",
       "3      ectoine/hydroxyectoine ABC transporter permeas...   \n",
       "4                                    YqzK family protein   \n",
       "...                                                  ...   \n",
       "49995   aminoglycoside phosphotransferase family protein   \n",
       "49996            response regulator transcription factor   \n",
       "49997                ABC transporter ATP-binding protein   \n",
       "49998                ABC transporter ATP-binding protein   \n",
       "49999            response regulator transcription factor   \n",
       "\n",
       "                                    t_protein_desc  m_protein_len  \\\n",
       "0      sigma-70 family RNA polymerase sigma factor            206   \n",
       "1          response regulator transcription factor            233   \n",
       "2                        SDR family oxidoreductase            287   \n",
       "3              amino acid ABC transporter permease            234   \n",
       "4                              YqzK family protein             80   \n",
       "...                                            ...            ...   \n",
       "49995              serine/threonine protein kinase            271   \n",
       "49996      response regulator transcription factor            200   \n",
       "49997          ABC transporter ATP-binding protein            331   \n",
       "49998          ABC transporter ATP-binding protein            238   \n",
       "49999                           response regulator            236   \n",
       "\n",
       "       t_protein_len  \n",
       "0                202  \n",
       "1                237  \n",
       "2                252  \n",
       "3                269  \n",
       "4                 66  \n",
       "...              ...  \n",
       "49995            353  \n",
       "49996            203  \n",
       "49997            338  \n",
       "49998            235  \n",
       "49999            211  \n",
       "\n",
       "[50000 rows x 30 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "c0562258",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.utils import resample\n",
    "\n",
    "# Assuming your data is in a pandas DataFrame called 'data'\n",
    "# Separate the majority and minority classes\n",
    "majority_class = target[target['protein_match'] == 'Yes']\n",
    "minority_class = target[target['protein_match'] == 'No']\n",
    "\n",
    "# Undersample the majority class to match the number of minority class samples\n",
    "n_samples = len(minority_class)\n",
    "undersampled_majority = resample(majority_class, n_samples=n_samples, replace=False)\n",
    "\n",
    "# Combine the undersampled majority class with the minority class\n",
    "balanced_data = pd.concat([undersampled_majority, minority_class])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "e929fd05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prot_pair_index</th>\n",
       "      <th>Jaccard_Score</th>\n",
       "      <th>protein_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4397</th>\n",
       "      <td>59874753</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>96832867</td>\n",
       "      <td>0.50</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>139598270</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>91444751</td>\n",
       "      <td>0.75</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5886</th>\n",
       "      <td>13192133</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6328</th>\n",
       "      <td>125335561</td>\n",
       "      <td>0.25</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6329</th>\n",
       "      <td>105441342</td>\n",
       "      <td>0.20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6336</th>\n",
       "      <td>87888523</td>\n",
       "      <td>0.20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6339</th>\n",
       "      <td>38582871</td>\n",
       "      <td>0.20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6347</th>\n",
       "      <td>90211763</td>\n",
       "      <td>0.20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2210 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      prot_pair_index  Jaccard_Score protein_match\n",
       "4397         59874753           1.00           Yes\n",
       "1205         96832867           0.50           Yes\n",
       "273         139598270           1.00           Yes\n",
       "208          91444751           0.75           Yes\n",
       "5886         13192133           1.00           Yes\n",
       "...               ...            ...           ...\n",
       "6328        125335561           0.25            No\n",
       "6329        105441342           0.20            No\n",
       "6336         87888523           0.20            No\n",
       "6339         38582871           0.20            No\n",
       "6347         90211763           0.20            No\n",
       "\n",
       "[2210 rows x 3 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balanced_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "792508bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2210, 32)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.merge(df, balanced_data, on=['prot_pair_index'])\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "aabaa4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "this list comes from a combination of reading through the features and determining which might be useful\n",
    "and timing some of the feature generations. those that took more than 30ish seconds were eliminated\n",
    "Also removed those that have really high dimensionality (>4000)\n",
    "\"\"\"\n",
    "\n",
    "feature_list = ['AAC', 'GAAC', 'DistancePair',\n",
    "               'CTDC', 'CTDT', 'CTDD', 'CTriad', 'GDPC type 1', 'GDPC type 2',\n",
    "                'CKSAAGP type 1', 'CKSAAGP type 2', 'PseKRAAC type 2', 'PseKRAAC type 3A','PseKRAAC type 7',\n",
    "                'PseKRAAC type 9', 'Geary','APAAC', 'QSOrder']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "10e7fe5d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# df = create_new_dataframe(df, 'seq_50k.fasta', descriptors= ['EGAAC', 'CKSAAGP type 1', 'CKSAAGP type 2',\n",
    "#                                                             'GDPC type 1', 'GDPC type 2', 'TPC type 1', 'TPC type 2',\n",
    "#                                                             'GTPC type 1', 'GTPC type 2', 'DPC type 1', 'DPC type 2',\n",
    "#                                                             'Moran', 'Geary', 'NMBroto', 'AC', 'CC', 'ACC',\n",
    "#                                                             'PAAC', 'APAAC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "78c6431f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File imported successfully.\n",
      "File imported successfully.\n"
     ]
    }
   ],
   "source": [
    "df = create_new_dataframe(df, ['seq_50k_a.fasta', 'seq_50k_b.fasta'], descriptors=['AAC', 'GAAC', 'QSOrder'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90871d01",
   "metadata": {},
   "source": [
    "Split data into dev and test, and then split that into train and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6519dcaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop columns that don't exihibit signficant pearson correlation with bit_score\n",
    "\n",
    "df = df.drop(columns = ['meso_index', 'meso_protein_int_index', 'local_gap_compressed_percent_id_16s', \n",
    "                        'scaled_local_query_percent_id_16s', 'scaled_local_symmetric_percent_id_16s',\n",
    "                       'bit_score_16s', 'm_ogt', 't_ogt', 'taxa_pair_index', 'thermo_protein_int_index'\n",
    "                       , 'prot_pair_index', 'ogt_difference', 'Jaccard_Score',\n",
    "                       'query_align_cov_16s', 'subject_align_cov_16s',\n",
    "                       'Unnamed: 0', 'thermo_index','m_protein_seq', 't_protein_seq', \n",
    "                       'm_protein_desc', 't_protein_desc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e571a475",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>...</th>\n",
       "      <th>QSOrder_Grantham.Xr.T</th>\n",
       "      <th>QSOrder_Grantham.Xr.W</th>\n",
       "      <th>QSOrder_Grantham.Xr.Y</th>\n",
       "      <th>QSOrder_Grantham.Xr.V</th>\n",
       "      <th>QSOrder_Schneider.Xd.1</th>\n",
       "      <th>QSOrder_Schneider.Xd.2</th>\n",
       "      <th>QSOrder_Schneider.Xd.3</th>\n",
       "      <th>QSOrder_Grantham.Xd.1</th>\n",
       "      <th>QSOrder_Grantham.Xd.2</th>\n",
       "      <th>QSOrder_Grantham.Xd.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.327273</td>\n",
       "      <td>0.200743</td>\n",
       "      <td>0.214712</td>\n",
       "      <td>166</td>\n",
       "      <td>0.617100</td>\n",
       "      <td>163</td>\n",
       "      <td>0.696581</td>\n",
       "      <td>175</td>\n",
       "      <td>234</td>\n",
       "      <td>269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.849359</td>\n",
       "      <td>2.807602</td>\n",
       "      <td>0.656323</td>\n",
       "      <td>1.323584</td>\n",
       "      <td>0.950250</td>\n",
       "      <td>1.057052</td>\n",
       "      <td>0.957343</td>\n",
       "      <td>0.921186</td>\n",
       "      <td>1.106142</td>\n",
       "      <td>0.987225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.840951</td>\n",
       "      <td>1.324497</td>\n",
       "      <td>0.378428</td>\n",
       "      <td>1.000131</td>\n",
       "      <td>1.035213</td>\n",
       "      <td>0.999638</td>\n",
       "      <td>1.027001</td>\n",
       "      <td>0.972217</td>\n",
       "      <td>1.003861</td>\n",
       "      <td>1.025202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.354067</td>\n",
       "      <td>0.245847</td>\n",
       "      <td>0.225954</td>\n",
       "      <td>208</td>\n",
       "      <td>0.691030</td>\n",
       "      <td>216</td>\n",
       "      <td>0.610170</td>\n",
       "      <td>264</td>\n",
       "      <td>354</td>\n",
       "      <td>301</td>\n",
       "      <td>...</td>\n",
       "      <td>1.199196</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.782085</td>\n",
       "      <td>0.857770</td>\n",
       "      <td>0.973425</td>\n",
       "      <td>1.016259</td>\n",
       "      <td>1.037198</td>\n",
       "      <td>0.967060</td>\n",
       "      <td>0.991860</td>\n",
       "      <td>1.044636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.331915</td>\n",
       "      <td>0.245283</td>\n",
       "      <td>0.242613</td>\n",
       "      <td>239</td>\n",
       "      <td>0.751572</td>\n",
       "      <td>238</td>\n",
       "      <td>0.732308</td>\n",
       "      <td>202</td>\n",
       "      <td>325</td>\n",
       "      <td>318</td>\n",
       "      <td>...</td>\n",
       "      <td>0.398850</td>\n",
       "      <td>0.864175</td>\n",
       "      <td>0.439943</td>\n",
       "      <td>1.639541</td>\n",
       "      <td>1.072840</td>\n",
       "      <td>0.907991</td>\n",
       "      <td>1.008479</td>\n",
       "      <td>1.081410</td>\n",
       "      <td>0.899625</td>\n",
       "      <td>1.023937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.322430</td>\n",
       "      <td>0.264368</td>\n",
       "      <td>0.245552</td>\n",
       "      <td>232</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>208</td>\n",
       "      <td>0.691030</td>\n",
       "      <td>227</td>\n",
       "      <td>301</td>\n",
       "      <td>261</td>\n",
       "      <td>...</td>\n",
       "      <td>1.878335</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.717419</td>\n",
       "      <td>1.067520</td>\n",
       "      <td>1.034354</td>\n",
       "      <td>0.945732</td>\n",
       "      <td>1.058166</td>\n",
       "      <td>0.979840</td>\n",
       "      <td>1.045301</td>\n",
       "      <td>0.977320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2205</th>\n",
       "      <td>0.290909</td>\n",
       "      <td>0.162437</td>\n",
       "      <td>0.154589</td>\n",
       "      <td>110</td>\n",
       "      <td>0.558376</td>\n",
       "      <td>112</td>\n",
       "      <td>0.516129</td>\n",
       "      <td>85</td>\n",
       "      <td>217</td>\n",
       "      <td>197</td>\n",
       "      <td>...</td>\n",
       "      <td>1.466560</td>\n",
       "      <td>0.879936</td>\n",
       "      <td>1.173248</td>\n",
       "      <td>1.613216</td>\n",
       "      <td>0.931433</td>\n",
       "      <td>1.004820</td>\n",
       "      <td>1.103878</td>\n",
       "      <td>0.981874</td>\n",
       "      <td>0.948476</td>\n",
       "      <td>1.068755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.488462</td>\n",
       "      <td>0.455197</td>\n",
       "      <td>258</td>\n",
       "      <td>0.992308</td>\n",
       "      <td>263</td>\n",
       "      <td>0.882550</td>\n",
       "      <td>619</td>\n",
       "      <td>298</td>\n",
       "      <td>260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.833497</td>\n",
       "      <td>0.983990</td>\n",
       "      <td>1.062709</td>\n",
       "      <td>0.695821</td>\n",
       "      <td>1.074279</td>\n",
       "      <td>0.985104</td>\n",
       "      <td>0.969361</td>\n",
       "      <td>1.053592</td>\n",
       "      <td>0.951543</td>\n",
       "      <td>1.000276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2207</th>\n",
       "      <td>0.383838</td>\n",
       "      <td>0.376238</td>\n",
       "      <td>0.374384</td>\n",
       "      <td>296</td>\n",
       "      <td>0.976898</td>\n",
       "      <td>299</td>\n",
       "      <td>0.977124</td>\n",
       "      <td>436</td>\n",
       "      <td>306</td>\n",
       "      <td>303</td>\n",
       "      <td>...</td>\n",
       "      <td>1.695522</td>\n",
       "      <td>0.726652</td>\n",
       "      <td>0.322957</td>\n",
       "      <td>0.523713</td>\n",
       "      <td>0.946281</td>\n",
       "      <td>0.992210</td>\n",
       "      <td>1.060345</td>\n",
       "      <td>0.938189</td>\n",
       "      <td>1.001573</td>\n",
       "      <td>1.063319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>0.769231</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>78</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>258</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375423</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.126270</td>\n",
       "      <td>1.126270</td>\n",
       "      <td>1.073105</td>\n",
       "      <td>1.012337</td>\n",
       "      <td>0.921517</td>\n",
       "      <td>1.023642</td>\n",
       "      <td>0.964341</td>\n",
       "      <td>1.017542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2209</th>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>251</td>\n",
       "      <td>362</td>\n",
       "      <td>...</td>\n",
       "      <td>0.730568</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.175617</td>\n",
       "      <td>0.842963</td>\n",
       "      <td>0.904099</td>\n",
       "      <td>1.050159</td>\n",
       "      <td>0.993395</td>\n",
       "      <td>1.024498</td>\n",
       "      <td>1.003472</td>\n",
       "      <td>0.973083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2210 rows × 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                            0.327273                       0.200743   \n",
       "1                            0.275281                       0.182156   \n",
       "2                            0.354067                       0.245847   \n",
       "3                            0.331915                       0.245283   \n",
       "4                            0.322430                       0.264368   \n",
       "...                               ...                            ...   \n",
       "2205                         0.290909                       0.162437   \n",
       "2206                         0.488462                       0.488462   \n",
       "2207                         0.383838                       0.376238   \n",
       "2208                         0.769231                       0.769231   \n",
       "2209                         0.343220                       0.223757   \n",
       "\n",
       "      scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                              0.214712              166         0.617100   \n",
       "1                              0.152174              185         0.687732   \n",
       "2                              0.225954              208         0.691030   \n",
       "3                              0.242613              239         0.751572   \n",
       "4                              0.245552              232         0.888889   \n",
       "...                                 ...              ...              ...   \n",
       "2205                           0.154589              110         0.558376   \n",
       "2206                           0.455197              258         0.992308   \n",
       "2207                           0.374384              296         0.976898   \n",
       "2208                           0.769231               78         1.000000   \n",
       "2209                           0.264274              234         0.646409   \n",
       "\n",
       "      subject_align_len  subject_align_cov  bit_score  m_protein_len  \\\n",
       "0                   163           0.696581        175            234   \n",
       "1                   197           0.525333         48            375   \n",
       "2                   216           0.610170        264            354   \n",
       "3                   238           0.732308        202            325   \n",
       "4                   208           0.691030        227            301   \n",
       "...                 ...                ...        ...            ...   \n",
       "2205                112           0.516129         85            217   \n",
       "2206                263           0.882550        619            298   \n",
       "2207                299           0.977124        436            306   \n",
       "2208                 78           1.000000        258             78   \n",
       "2209                238           0.948207        347            251   \n",
       "\n",
       "      t_protein_len  ... QSOrder_Grantham.Xr.T  QSOrder_Grantham.Xr.W  \\\n",
       "0               269  ...              0.849359               2.807602   \n",
       "1               269  ...              0.840951               1.324497   \n",
       "2               301  ...              1.199196                    inf   \n",
       "3               318  ...              0.398850               0.864175   \n",
       "4               261  ...              1.878335                    inf   \n",
       "...             ...  ...                   ...                    ...   \n",
       "2205            197  ...              1.466560               0.879936   \n",
       "2206            260  ...              0.833497               0.983990   \n",
       "2207            303  ...              1.695522               0.726652   \n",
       "2208             78  ...              0.375423                    NaN   \n",
       "2209            362  ...              0.730568               0.000000   \n",
       "\n",
       "      QSOrder_Grantham.Xr.Y  QSOrder_Grantham.Xr.V  QSOrder_Schneider.Xd.1  \\\n",
       "0                  0.656323               1.323584                0.950250   \n",
       "1                  0.378428               1.000131                1.035213   \n",
       "2                  0.782085               0.857770                0.973425   \n",
       "3                  0.439943               1.639541                1.072840   \n",
       "4                  0.717419               1.067520                1.034354   \n",
       "...                     ...                    ...                     ...   \n",
       "2205               1.173248               1.613216                0.931433   \n",
       "2206               1.062709               0.695821                1.074279   \n",
       "2207               0.322957               0.523713                0.946281   \n",
       "2208               1.126270               1.126270                1.073105   \n",
       "2209               0.175617               0.842963                0.904099   \n",
       "\n",
       "      QSOrder_Schneider.Xd.2  QSOrder_Schneider.Xd.3  QSOrder_Grantham.Xd.1  \\\n",
       "0                   1.057052                0.957343               0.921186   \n",
       "1                   0.999638                1.027001               0.972217   \n",
       "2                   1.016259                1.037198               0.967060   \n",
       "3                   0.907991                1.008479               1.081410   \n",
       "4                   0.945732                1.058166               0.979840   \n",
       "...                      ...                     ...                    ...   \n",
       "2205                1.004820                1.103878               0.981874   \n",
       "2206                0.985104                0.969361               1.053592   \n",
       "2207                0.992210                1.060345               0.938189   \n",
       "2208                1.012337                0.921517               1.023642   \n",
       "2209                1.050159                0.993395               1.024498   \n",
       "\n",
       "      QSOrder_Grantham.Xd.2  QSOrder_Grantham.Xd.3  \n",
       "0                  1.106142               0.987225  \n",
       "1                  1.003861               1.025202  \n",
       "2                  0.991860               1.044636  \n",
       "3                  0.899625               1.023937  \n",
       "4                  1.045301               0.977320  \n",
       "...                     ...                    ...  \n",
       "2205               0.948476               1.068755  \n",
       "2206               0.951543               1.000276  \n",
       "2207               1.001573               1.063319  \n",
       "2208               0.964341               1.017542  \n",
       "2209               1.003472               0.973083  \n",
       "\n",
       "[2210 rows x 82 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=df.drop(columns=['index_y', 'index_x'])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "1e45ecac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
       "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
       "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
       "       'bit_score', 'm_protein_len', 't_protein_len', 'protein_match', 'AAC_A',\n",
       "       'AAC_C', 'AAC_D', 'AAC_E', 'AAC_F', 'AAC_G', 'AAC_H', 'AAC_I', 'AAC_K',\n",
       "       'AAC_L', 'AAC_M', 'AAC_N', 'AAC_P', 'AAC_Q', 'AAC_R', 'AAC_S', 'AAC_T',\n",
       "       'AAC_V', 'AAC_W', 'AAC_Y', 'GAAC_alphatic', 'GAAC_aromatic',\n",
       "       'GAAC_postivecharge', 'GAAC_negativecharge', 'GAAC_uncharge',\n",
       "       'QSOrder_Schneider.Xr.A', 'QSOrder_Schneider.Xr.R',\n",
       "       'QSOrder_Schneider.Xr.N', 'QSOrder_Schneider.Xr.D',\n",
       "       'QSOrder_Schneider.Xr.C', 'QSOrder_Schneider.Xr.Q',\n",
       "       'QSOrder_Schneider.Xr.E', 'QSOrder_Schneider.Xr.G',\n",
       "       'QSOrder_Schneider.Xr.H', 'QSOrder_Schneider.Xr.I',\n",
       "       'QSOrder_Schneider.Xr.L', 'QSOrder_Schneider.Xr.K',\n",
       "       'QSOrder_Schneider.Xr.M', 'QSOrder_Schneider.Xr.F',\n",
       "       'QSOrder_Schneider.Xr.P', 'QSOrder_Schneider.Xr.S',\n",
       "       'QSOrder_Schneider.Xr.T', 'QSOrder_Schneider.Xr.W',\n",
       "       'QSOrder_Schneider.Xr.Y', 'QSOrder_Schneider.Xr.V',\n",
       "       'QSOrder_Grantham.Xr.A', 'QSOrder_Grantham.Xr.R',\n",
       "       'QSOrder_Grantham.Xr.N', 'QSOrder_Grantham.Xr.D',\n",
       "       'QSOrder_Grantham.Xr.C', 'QSOrder_Grantham.Xr.Q',\n",
       "       'QSOrder_Grantham.Xr.E', 'QSOrder_Grantham.Xr.G',\n",
       "       'QSOrder_Grantham.Xr.H', 'QSOrder_Grantham.Xr.I',\n",
       "       'QSOrder_Grantham.Xr.L', 'QSOrder_Grantham.Xr.K',\n",
       "       'QSOrder_Grantham.Xr.M', 'QSOrder_Grantham.Xr.F',\n",
       "       'QSOrder_Grantham.Xr.P', 'QSOrder_Grantham.Xr.S',\n",
       "       'QSOrder_Grantham.Xr.T', 'QSOrder_Grantham.Xr.W',\n",
       "       'QSOrder_Grantham.Xr.Y', 'QSOrder_Grantham.Xr.V',\n",
       "       'QSOrder_Schneider.Xd.1', 'QSOrder_Schneider.Xd.2',\n",
       "       'QSOrder_Schneider.Xd.3', 'QSOrder_Grantham.Xd.1',\n",
       "       'QSOrder_Grantham.Xd.2', 'QSOrder_Grantham.Xd.3'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "29ba7c2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=df.drop(columns=['local_gap_compressed_percent_id', 'scaled_local_query_percent_id',\n",
    "       'scaled_local_symmetric_percent_id', 'query_align_len',\n",
    "       'query_align_cov', 'subject_align_len', 'subject_align_cov',\n",
    "       'bit_score', 'm_protein_len', 't_protein_len'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "3d166740",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_match</th>\n",
       "      <th>AAC_A</th>\n",
       "      <th>AAC_C</th>\n",
       "      <th>AAC_D</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>AAC_F</th>\n",
       "      <th>AAC_G</th>\n",
       "      <th>AAC_H</th>\n",
       "      <th>AAC_I</th>\n",
       "      <th>AAC_K</th>\n",
       "      <th>...</th>\n",
       "      <th>QSOrder_Grantham.Xr.T</th>\n",
       "      <th>QSOrder_Grantham.Xr.W</th>\n",
       "      <th>QSOrder_Grantham.Xr.Y</th>\n",
       "      <th>QSOrder_Grantham.Xr.V</th>\n",
       "      <th>QSOrder_Schneider.Xd.1</th>\n",
       "      <th>QSOrder_Schneider.Xd.2</th>\n",
       "      <th>QSOrder_Schneider.Xd.3</th>\n",
       "      <th>QSOrder_Grantham.Xd.1</th>\n",
       "      <th>QSOrder_Grantham.Xd.2</th>\n",
       "      <th>QSOrder_Grantham.Xd.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Yes</td>\n",
       "      <td>0.008134</td>\n",
       "      <td>-0.003717</td>\n",
       "      <td>-0.016919</td>\n",
       "      <td>0.005560</td>\n",
       "      <td>0.010946</td>\n",
       "      <td>0.015950</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>-0.018238</td>\n",
       "      <td>0.008722</td>\n",
       "      <td>...</td>\n",
       "      <td>0.849359</td>\n",
       "      <td>2.807602</td>\n",
       "      <td>0.656323</td>\n",
       "      <td>1.323584</td>\n",
       "      <td>0.950250</td>\n",
       "      <td>1.057052</td>\n",
       "      <td>0.957343</td>\n",
       "      <td>0.921186</td>\n",
       "      <td>1.106142</td>\n",
       "      <td>0.987225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>No</td>\n",
       "      <td>0.074577</td>\n",
       "      <td>-0.010587</td>\n",
       "      <td>0.021006</td>\n",
       "      <td>-0.009864</td>\n",
       "      <td>0.005978</td>\n",
       "      <td>-0.005988</td>\n",
       "      <td>-0.005740</td>\n",
       "      <td>-0.004610</td>\n",
       "      <td>-0.008486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.840951</td>\n",
       "      <td>1.324497</td>\n",
       "      <td>0.378428</td>\n",
       "      <td>1.000131</td>\n",
       "      <td>1.035213</td>\n",
       "      <td>0.999638</td>\n",
       "      <td>1.027001</td>\n",
       "      <td>0.972217</td>\n",
       "      <td>1.003861</td>\n",
       "      <td>1.025202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.055624</td>\n",
       "      <td>0.010802</td>\n",
       "      <td>0.022618</td>\n",
       "      <td>-0.007621</td>\n",
       "      <td>0.013965</td>\n",
       "      <td>-0.016255</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.037061</td>\n",
       "      <td>0.030079</td>\n",
       "      <td>...</td>\n",
       "      <td>1.199196</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.782085</td>\n",
       "      <td>0.857770</td>\n",
       "      <td>0.973425</td>\n",
       "      <td>1.016259</td>\n",
       "      <td>1.037198</td>\n",
       "      <td>0.967060</td>\n",
       "      <td>0.991860</td>\n",
       "      <td>1.044636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.021239</td>\n",
       "      <td>0.003077</td>\n",
       "      <td>0.008824</td>\n",
       "      <td>-0.006628</td>\n",
       "      <td>0.023599</td>\n",
       "      <td>0.007876</td>\n",
       "      <td>-0.015723</td>\n",
       "      <td>0.014233</td>\n",
       "      <td>0.005883</td>\n",
       "      <td>...</td>\n",
       "      <td>0.398850</td>\n",
       "      <td>0.864175</td>\n",
       "      <td>0.439943</td>\n",
       "      <td>1.639541</td>\n",
       "      <td>1.072840</td>\n",
       "      <td>0.907991</td>\n",
       "      <td>1.008479</td>\n",
       "      <td>1.081410</td>\n",
       "      <td>0.899625</td>\n",
       "      <td>1.023937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.009139</td>\n",
       "      <td>0.002813</td>\n",
       "      <td>0.004366</td>\n",
       "      <td>-0.009165</td>\n",
       "      <td>-0.000242</td>\n",
       "      <td>0.007205</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>-0.022721</td>\n",
       "      <td>-0.028347</td>\n",
       "      <td>...</td>\n",
       "      <td>1.878335</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.717419</td>\n",
       "      <td>1.067520</td>\n",
       "      <td>1.034354</td>\n",
       "      <td>0.945732</td>\n",
       "      <td>1.058166</td>\n",
       "      <td>0.979840</td>\n",
       "      <td>1.045301</td>\n",
       "      <td>0.977320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2205</th>\n",
       "      <td>No</td>\n",
       "      <td>-0.069803</td>\n",
       "      <td>0.004608</td>\n",
       "      <td>0.022036</td>\n",
       "      <td>0.016959</td>\n",
       "      <td>0.012889</td>\n",
       "      <td>-0.022246</td>\n",
       "      <td>-0.020772</td>\n",
       "      <td>-0.033264</td>\n",
       "      <td>0.017030</td>\n",
       "      <td>...</td>\n",
       "      <td>1.466560</td>\n",
       "      <td>0.879936</td>\n",
       "      <td>1.173248</td>\n",
       "      <td>1.613216</td>\n",
       "      <td>0.931433</td>\n",
       "      <td>1.004820</td>\n",
       "      <td>1.103878</td>\n",
       "      <td>0.981874</td>\n",
       "      <td>0.948476</td>\n",
       "      <td>1.068755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.013320</td>\n",
       "      <td>-0.000981</td>\n",
       "      <td>0.001652</td>\n",
       "      <td>0.019979</td>\n",
       "      <td>0.011461</td>\n",
       "      <td>0.005498</td>\n",
       "      <td>0.001394</td>\n",
       "      <td>-0.005885</td>\n",
       "      <td>-0.001962</td>\n",
       "      <td>...</td>\n",
       "      <td>0.833497</td>\n",
       "      <td>0.983990</td>\n",
       "      <td>1.062709</td>\n",
       "      <td>0.695821</td>\n",
       "      <td>1.074279</td>\n",
       "      <td>0.985104</td>\n",
       "      <td>0.969361</td>\n",
       "      <td>1.053592</td>\n",
       "      <td>0.951543</td>\n",
       "      <td>1.000276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2207</th>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.004821</td>\n",
       "      <td>-0.000032</td>\n",
       "      <td>-0.000453</td>\n",
       "      <td>0.002815</td>\n",
       "      <td>-0.000356</td>\n",
       "      <td>-0.013881</td>\n",
       "      <td>0.019252</td>\n",
       "      <td>0.035462</td>\n",
       "      <td>0.009707</td>\n",
       "      <td>...</td>\n",
       "      <td>1.695522</td>\n",
       "      <td>0.726652</td>\n",
       "      <td>0.322957</td>\n",
       "      <td>0.523713</td>\n",
       "      <td>0.946281</td>\n",
       "      <td>0.992210</td>\n",
       "      <td>1.060345</td>\n",
       "      <td>0.938189</td>\n",
       "      <td>1.001573</td>\n",
       "      <td>1.063319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>Yes</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>-0.038462</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.012821</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>...</td>\n",
       "      <td>0.375423</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.126270</td>\n",
       "      <td>1.126270</td>\n",
       "      <td>1.073105</td>\n",
       "      <td>1.012337</td>\n",
       "      <td>0.921517</td>\n",
       "      <td>1.023642</td>\n",
       "      <td>0.964341</td>\n",
       "      <td>1.017542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2209</th>\n",
       "      <td>No</td>\n",
       "      <td>0.012745</td>\n",
       "      <td>-0.004303</td>\n",
       "      <td>-0.001013</td>\n",
       "      <td>0.002972</td>\n",
       "      <td>0.008232</td>\n",
       "      <td>-0.014241</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.003610</td>\n",
       "      <td>0.050835</td>\n",
       "      <td>...</td>\n",
       "      <td>0.730568</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.175617</td>\n",
       "      <td>0.842963</td>\n",
       "      <td>0.904099</td>\n",
       "      <td>1.050159</td>\n",
       "      <td>0.993395</td>\n",
       "      <td>1.024498</td>\n",
       "      <td>1.003472</td>\n",
       "      <td>0.973083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2210 rows × 72 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     protein_match     AAC_A     AAC_C     AAC_D     AAC_E     AAC_F  \\\n",
       "0              Yes  0.008134 -0.003717 -0.016919  0.005560  0.010946   \n",
       "1               No  0.074577 -0.010587  0.021006 -0.009864  0.005978   \n",
       "2              Yes -0.055624  0.010802  0.022618 -0.007621  0.013965   \n",
       "3              Yes -0.021239  0.003077  0.008824 -0.006628  0.023599   \n",
       "4              Yes -0.009139  0.002813  0.004366 -0.009165 -0.000242   \n",
       "...            ...       ...       ...       ...       ...       ...   \n",
       "2205            No -0.069803  0.004608  0.022036  0.016959  0.012889   \n",
       "2206           Yes -0.013320 -0.000981  0.001652  0.019979  0.011461   \n",
       "2207           Yes -0.004821 -0.000032 -0.000453  0.002815 -0.000356   \n",
       "2208           Yes  0.038462 -0.038462 -0.012821 -0.012821 -0.012821   \n",
       "2209            No  0.012745 -0.004303 -0.001013  0.002972  0.008232   \n",
       "\n",
       "         AAC_G     AAC_H     AAC_I     AAC_K  ...  QSOrder_Grantham.Xr.T  \\\n",
       "0     0.015950  0.004274 -0.018238  0.008722  ...               0.849359   \n",
       "1    -0.005988 -0.005740 -0.004610 -0.008486  ...               0.840951   \n",
       "2    -0.016255  0.000338  0.037061  0.030079  ...               1.199196   \n",
       "3     0.007876 -0.015723  0.014233  0.005883  ...               0.398850   \n",
       "4     0.007205  0.000267 -0.022721 -0.028347  ...               1.878335   \n",
       "...        ...       ...       ...       ...  ...                    ...   \n",
       "2205 -0.022246 -0.020772 -0.033264  0.017030  ...               1.466560   \n",
       "2206  0.005498  0.001394 -0.005885 -0.001962  ...               0.833497   \n",
       "2207 -0.013881  0.019252  0.035462  0.009707  ...               1.695522   \n",
       "2208  0.000000 -0.012821  0.000000  0.025641  ...               0.375423   \n",
       "2209 -0.014241  0.005789  0.003610  0.050835  ...               0.730568   \n",
       "\n",
       "      QSOrder_Grantham.Xr.W  QSOrder_Grantham.Xr.Y  QSOrder_Grantham.Xr.V  \\\n",
       "0                  2.807602               0.656323               1.323584   \n",
       "1                  1.324497               0.378428               1.000131   \n",
       "2                       inf               0.782085               0.857770   \n",
       "3                  0.864175               0.439943               1.639541   \n",
       "4                       inf               0.717419               1.067520   \n",
       "...                     ...                    ...                    ...   \n",
       "2205               0.879936               1.173248               1.613216   \n",
       "2206               0.983990               1.062709               0.695821   \n",
       "2207               0.726652               0.322957               0.523713   \n",
       "2208                    NaN               1.126270               1.126270   \n",
       "2209               0.000000               0.175617               0.842963   \n",
       "\n",
       "      QSOrder_Schneider.Xd.1  QSOrder_Schneider.Xd.2  QSOrder_Schneider.Xd.3  \\\n",
       "0                   0.950250                1.057052                0.957343   \n",
       "1                   1.035213                0.999638                1.027001   \n",
       "2                   0.973425                1.016259                1.037198   \n",
       "3                   1.072840                0.907991                1.008479   \n",
       "4                   1.034354                0.945732                1.058166   \n",
       "...                      ...                     ...                     ...   \n",
       "2205                0.931433                1.004820                1.103878   \n",
       "2206                1.074279                0.985104                0.969361   \n",
       "2207                0.946281                0.992210                1.060345   \n",
       "2208                1.073105                1.012337                0.921517   \n",
       "2209                0.904099                1.050159                0.993395   \n",
       "\n",
       "      QSOrder_Grantham.Xd.1  QSOrder_Grantham.Xd.2  QSOrder_Grantham.Xd.3  \n",
       "0                  0.921186               1.106142               0.987225  \n",
       "1                  0.972217               1.003861               1.025202  \n",
       "2                  0.967060               0.991860               1.044636  \n",
       "3                  1.081410               0.899625               1.023937  \n",
       "4                  0.979840               1.045301               0.977320  \n",
       "...                     ...                    ...                    ...  \n",
       "2205               0.981874               0.948476               1.068755  \n",
       "2206               1.053592               0.951543               1.000276  \n",
       "2207               0.938189               1.001573               1.063319  \n",
       "2208               1.023642               0.964341               1.017542  \n",
       "2209               1.024498               1.003472               0.973083  \n",
       "\n",
       "[2210 rows x 72 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "72997b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "#maybe divide all of the AAC/GAAC columns by the protein length "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "63777c88",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2210, 72)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6894895e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.replace([np.inf, -np.inf], np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "bd0163ce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein_match             0\n",
      "AAC_A                     0\n",
      "AAC_C                     0\n",
      "AAC_D                     0\n",
      "AAC_E                     0\n",
      "AAC_F                     0\n",
      "AAC_G                     0\n",
      "AAC_H                     0\n",
      "AAC_I                     0\n",
      "AAC_K                     0\n",
      "AAC_L                     0\n",
      "AAC_M                     0\n",
      "AAC_N                     0\n",
      "AAC_P                     0\n",
      "AAC_Q                     0\n",
      "AAC_R                     0\n",
      "AAC_S                     0\n",
      "AAC_T                     0\n",
      "AAC_V                     0\n",
      "AAC_W                     0\n",
      "AAC_Y                     0\n",
      "GAAC_alphatic             0\n",
      "GAAC_aromatic             0\n",
      "GAAC_postivecharge        0\n",
      "GAAC_negativecharge       0\n",
      "GAAC_uncharge             0\n",
      "QSOrder_Schneider.Xr.A    0\n",
      "QSOrder_Schneider.Xr.S    0\n",
      "QSOrder_Schneider.Xr.V    0\n",
      "QSOrder_Grantham.Xr.A     0\n",
      "QSOrder_Grantham.Xr.S     0\n",
      "QSOrder_Grantham.Xr.V     0\n",
      "QSOrder_Schneider.Xd.1    0\n",
      "QSOrder_Schneider.Xd.2    0\n",
      "QSOrder_Schneider.Xd.3    0\n",
      "QSOrder_Grantham.Xd.1     0\n",
      "QSOrder_Grantham.Xd.2     0\n",
      "QSOrder_Grantham.Xd.3     0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0])"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nan_counts = df.isna().sum()\n",
    "print(nan_counts)\n",
    "nan_counts.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "f791d8fc",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['QSOrder_Grantham.Xd.2', 'QSOrder_Grantham.Xd.3'], dtype='object')"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns[36:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "aa2aefa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "log_cols = df.columns[36:]\n",
    "df[log_cols] = df[log_cols].apply(lambda x: np.log(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "76ebfd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.dropna(axis=1, how='any')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "1cd365e1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_match             0\n",
       "AAC_A                     0\n",
       "AAC_C                     0\n",
       "AAC_D                     0\n",
       "AAC_E                     0\n",
       "                         ..\n",
       "QSOrder_Schneider.Xd.2    0\n",
       "QSOrder_Schneider.Xd.3    0\n",
       "QSOrder_Grantham.Xd.1     0\n",
       "QSOrder_Grantham.Xd.2     0\n",
       "QSOrder_Grantham.Xd.3     0\n",
       "Length: 72, dtype: int64"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nan_counts = df.isna().sum()\n",
    "nan_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e42e3144",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2210, 56)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbe1b7e",
   "metadata": {},
   "source": [
    "Use MRMR to select for the best features. Going to start by grouping into different categories of features generated from iFeature Omega."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5110f375",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 48.64it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['QSOrder_Schneider.Xr.A',\n",
       " 'QSOrder_Grantham.Xd.3',\n",
       " 'QSOrder_Grantham.Xr.R',\n",
       " 'QSOrder_Schneider.Xd.1',\n",
       " 'AAC_A',\n",
       " 'QSOrder_Schneider.Xd.3',\n",
       " 'QSOrder_Grantham.Xr.A',\n",
       " 'AAC_I',\n",
       " 'QSOrder_Schneider.Xr.R',\n",
       " 'AAC_S',\n",
       " 'GAAC_alphatic',\n",
       " 'AAC_E',\n",
       " 'QSOrder_Grantham.Xd.2',\n",
       " 'AAC_N',\n",
       " 'GAAC_negativecharge',\n",
       " 'QSOrder_Schneider.Xr.L',\n",
       " 'QSOrder_Schneider.Xr.G',\n",
       " 'AAC_T',\n",
       " 'GAAC_uncharge',\n",
       " 'AAC_R']"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#use MRMR to select for the best features from PseKRAAC\n",
    "df_subset = df.loc[:, df.columns != 'protein_match']\n",
    "print(type(df_subset))\n",
    "\n",
    "# select top 10 features using mRMR\n",
    "from mrmr import mrmr_classif\n",
    "selected_features = mrmr_classif(X=df_subset.iloc[:,10:], y=df['protein_match'], K=20)\n",
    "\n",
    "selected_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bef98034",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_gap_compressed_percent_id</th>\n",
       "      <th>scaled_local_query_percent_id</th>\n",
       "      <th>scaled_local_symmetric_percent_id</th>\n",
       "      <th>query_align_len</th>\n",
       "      <th>query_align_cov</th>\n",
       "      <th>subject_align_len</th>\n",
       "      <th>subject_align_cov</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>m_protein_len</th>\n",
       "      <th>t_protein_len</th>\n",
       "      <th>...</th>\n",
       "      <th>GAAC_alphatic</th>\n",
       "      <th>AAC_E</th>\n",
       "      <th>QSOrder_Grantham.Xd.2</th>\n",
       "      <th>AAC_N</th>\n",
       "      <th>GAAC_negativecharge</th>\n",
       "      <th>QSOrder_Schneider.Xr.L</th>\n",
       "      <th>QSOrder_Schneider.Xr.G</th>\n",
       "      <th>AAC_T</th>\n",
       "      <th>GAAC_uncharge</th>\n",
       "      <th>AAC_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.283582</td>\n",
       "      <td>0.231707</td>\n",
       "      <td>0.240506</td>\n",
       "      <td>133</td>\n",
       "      <td>0.810976</td>\n",
       "      <td>134</td>\n",
       "      <td>0.881579</td>\n",
       "      <td>99</td>\n",
       "      <td>152</td>\n",
       "      <td>164</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.015083</td>\n",
       "      <td>-0.013479</td>\n",
       "      <td>1.036564</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.004974</td>\n",
       "      <td>1.329084</td>\n",
       "      <td>0.701461</td>\n",
       "      <td>0.003851</td>\n",
       "      <td>0.059050</td>\n",
       "      <td>-0.034981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.268482</td>\n",
       "      <td>0.190083</td>\n",
       "      <td>0.189300</td>\n",
       "      <td>248</td>\n",
       "      <td>0.683196</td>\n",
       "      <td>271</td>\n",
       "      <td>0.740437</td>\n",
       "      <td>132</td>\n",
       "      <td>366</td>\n",
       "      <td>363</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078422</td>\n",
       "      <td>-0.030461</td>\n",
       "      <td>0.982711</td>\n",
       "      <td>0.002394</td>\n",
       "      <td>-0.006345</td>\n",
       "      <td>0.920360</td>\n",
       "      <td>1.465759</td>\n",
       "      <td>0.010432</td>\n",
       "      <td>-0.035113</td>\n",
       "      <td>-0.030800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.265306</td>\n",
       "      <td>0.201550</td>\n",
       "      <td>0.210243</td>\n",
       "      <td>320</td>\n",
       "      <td>0.826873</td>\n",
       "      <td>294</td>\n",
       "      <td>0.828169</td>\n",
       "      <td>140</td>\n",
       "      <td>355</td>\n",
       "      <td>387</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030913</td>\n",
       "      <td>-0.020017</td>\n",
       "      <td>0.978314</td>\n",
       "      <td>0.011479</td>\n",
       "      <td>-0.021691</td>\n",
       "      <td>0.886583</td>\n",
       "      <td>1.065708</td>\n",
       "      <td>0.018277</td>\n",
       "      <td>0.057837</td>\n",
       "      <td>-0.010125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.275281</td>\n",
       "      <td>0.182156</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>185</td>\n",
       "      <td>0.687732</td>\n",
       "      <td>197</td>\n",
       "      <td>0.525333</td>\n",
       "      <td>48</td>\n",
       "      <td>375</td>\n",
       "      <td>269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.064287</td>\n",
       "      <td>-0.009864</td>\n",
       "      <td>1.003861</td>\n",
       "      <td>0.005898</td>\n",
       "      <td>0.011143</td>\n",
       "      <td>1.139824</td>\n",
       "      <td>0.931587</td>\n",
       "      <td>-0.013581</td>\n",
       "      <td>-0.068788</td>\n",
       "      <td>0.016882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.394495</td>\n",
       "      <td>0.346774</td>\n",
       "      <td>0.369099</td>\n",
       "      <td>108</td>\n",
       "      <td>0.870968</td>\n",
       "      <td>106</td>\n",
       "      <td>0.972477</td>\n",
       "      <td>130</td>\n",
       "      <td>109</td>\n",
       "      <td>124</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.028781</td>\n",
       "      <td>-0.030334</td>\n",
       "      <td>0.917208</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.007547</td>\n",
       "      <td>1.961181</td>\n",
       "      <td>1.123222</td>\n",
       "      <td>0.013613</td>\n",
       "      <td>0.063332</td>\n",
       "      <td>-0.031444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2205</th>\n",
       "      <td>0.241758</td>\n",
       "      <td>0.134557</td>\n",
       "      <td>0.144975</td>\n",
       "      <td>181</td>\n",
       "      <td>0.553517</td>\n",
       "      <td>190</td>\n",
       "      <td>0.678571</td>\n",
       "      <td>107</td>\n",
       "      <td>280</td>\n",
       "      <td>327</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049945</td>\n",
       "      <td>0.007678</td>\n",
       "      <td>1.001902</td>\n",
       "      <td>0.024509</td>\n",
       "      <td>0.004642</td>\n",
       "      <td>0.949185</td>\n",
       "      <td>1.416965</td>\n",
       "      <td>-0.051409</td>\n",
       "      <td>-0.086435</td>\n",
       "      <td>-0.043283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.186391</td>\n",
       "      <td>0.223801</td>\n",
       "      <td>197</td>\n",
       "      <td>0.582840</td>\n",
       "      <td>192</td>\n",
       "      <td>0.853333</td>\n",
       "      <td>195</td>\n",
       "      <td>225</td>\n",
       "      <td>338</td>\n",
       "      <td>...</td>\n",
       "      <td>0.082104</td>\n",
       "      <td>-0.038396</td>\n",
       "      <td>0.931608</td>\n",
       "      <td>0.002998</td>\n",
       "      <td>-0.030901</td>\n",
       "      <td>1.624165</td>\n",
       "      <td>1.373086</td>\n",
       "      <td>0.014872</td>\n",
       "      <td>-0.030743</td>\n",
       "      <td>-0.005786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2207</th>\n",
       "      <td>0.290909</td>\n",
       "      <td>0.162437</td>\n",
       "      <td>0.154589</td>\n",
       "      <td>110</td>\n",
       "      <td>0.558376</td>\n",
       "      <td>112</td>\n",
       "      <td>0.516129</td>\n",
       "      <td>85</td>\n",
       "      <td>217</td>\n",
       "      <td>197</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.074037</td>\n",
       "      <td>0.016959</td>\n",
       "      <td>0.948476</td>\n",
       "      <td>0.008281</td>\n",
       "      <td>0.038995</td>\n",
       "      <td>0.890999</td>\n",
       "      <td>0.712799</td>\n",
       "      <td>0.015626</td>\n",
       "      <td>0.022831</td>\n",
       "      <td>0.000327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>0.447368</td>\n",
       "      <td>0.387833</td>\n",
       "      <td>0.373626</td>\n",
       "      <td>229</td>\n",
       "      <td>0.870722</td>\n",
       "      <td>233</td>\n",
       "      <td>0.823322</td>\n",
       "      <td>448</td>\n",
       "      <td>283</td>\n",
       "      <td>263</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.037660</td>\n",
       "      <td>-0.007833</td>\n",
       "      <td>0.961154</td>\n",
       "      <td>-0.009217</td>\n",
       "      <td>0.006610</td>\n",
       "      <td>0.900462</td>\n",
       "      <td>0.840163</td>\n",
       "      <td>-0.021430</td>\n",
       "      <td>0.020557</td>\n",
       "      <td>0.004958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2209</th>\n",
       "      <td>0.343220</td>\n",
       "      <td>0.223757</td>\n",
       "      <td>0.264274</td>\n",
       "      <td>234</td>\n",
       "      <td>0.646409</td>\n",
       "      <td>238</td>\n",
       "      <td>0.948207</td>\n",
       "      <td>347</td>\n",
       "      <td>251</td>\n",
       "      <td>362</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012778</td>\n",
       "      <td>0.002972</td>\n",
       "      <td>1.003472</td>\n",
       "      <td>0.012535</td>\n",
       "      <td>0.001959</td>\n",
       "      <td>0.970531</td>\n",
       "      <td>0.822097</td>\n",
       "      <td>-0.017268</td>\n",
       "      <td>0.003654</td>\n",
       "      <td>-0.062369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2210 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      local_gap_compressed_percent_id  scaled_local_query_percent_id  \\\n",
       "0                            0.283582                       0.231707   \n",
       "1                            0.268482                       0.190083   \n",
       "2                            0.265306                       0.201550   \n",
       "3                            0.275281                       0.182156   \n",
       "4                            0.394495                       0.346774   \n",
       "...                               ...                            ...   \n",
       "2205                         0.241758                       0.134557   \n",
       "2206                         0.321429                       0.186391   \n",
       "2207                         0.290909                       0.162437   \n",
       "2208                         0.447368                       0.387833   \n",
       "2209                         0.343220                       0.223757   \n",
       "\n",
       "      scaled_local_symmetric_percent_id  query_align_len  query_align_cov  \\\n",
       "0                              0.240506              133         0.810976   \n",
       "1                              0.189300              248         0.683196   \n",
       "2                              0.210243              320         0.826873   \n",
       "3                              0.152174              185         0.687732   \n",
       "4                              0.369099              108         0.870968   \n",
       "...                                 ...              ...              ...   \n",
       "2205                           0.144975              181         0.553517   \n",
       "2206                           0.223801              197         0.582840   \n",
       "2207                           0.154589              110         0.558376   \n",
       "2208                           0.373626              229         0.870722   \n",
       "2209                           0.264274              234         0.646409   \n",
       "\n",
       "      subject_align_len  subject_align_cov  bit_score  m_protein_len  \\\n",
       "0                   134           0.881579         99            152   \n",
       "1                   271           0.740437        132            366   \n",
       "2                   294           0.828169        140            355   \n",
       "3                   197           0.525333         48            375   \n",
       "4                   106           0.972477        130            109   \n",
       "...                 ...                ...        ...            ...   \n",
       "2205                190           0.678571        107            280   \n",
       "2206                192           0.853333        195            225   \n",
       "2207                112           0.516129         85            217   \n",
       "2208                233           0.823322        448            283   \n",
       "2209                238           0.948207        347            251   \n",
       "\n",
       "      t_protein_len  ... GAAC_alphatic     AAC_E  QSOrder_Grantham.Xd.2  \\\n",
       "0               164  ...     -0.015083 -0.013479               1.036564   \n",
       "1               363  ...      0.078422 -0.030461               0.982711   \n",
       "2               387  ...      0.030913 -0.020017               0.978314   \n",
       "3               269  ...      0.064287 -0.009864               1.003861   \n",
       "4               124  ...     -0.028781 -0.030334               0.917208   \n",
       "...             ...  ...           ...       ...                    ...   \n",
       "2205            327  ...      0.049945  0.007678               1.001902   \n",
       "2206            338  ...      0.082104 -0.038396               0.931608   \n",
       "2207            197  ...     -0.074037  0.016959               0.948476   \n",
       "2208            263  ...     -0.037660 -0.007833               0.961154   \n",
       "2209            362  ...      0.012778  0.002972               1.003472   \n",
       "\n",
       "         AAC_N  GAAC_negativecharge  QSOrder_Schneider.Xr.L  \\\n",
       "0     0.006579             0.004974                1.329084   \n",
       "1     0.002394            -0.006345                0.920360   \n",
       "2     0.011479            -0.021691                0.886583   \n",
       "3     0.005898             0.011143                1.139824   \n",
       "4     0.000000            -0.007547                1.961181   \n",
       "...        ...                  ...                     ...   \n",
       "2205  0.024509             0.004642                0.949185   \n",
       "2206  0.002998            -0.030901                1.624165   \n",
       "2207  0.008281             0.038995                0.890999   \n",
       "2208 -0.009217             0.006610                0.900462   \n",
       "2209  0.012535             0.001959                0.970531   \n",
       "\n",
       "      QSOrder_Schneider.Xr.G     AAC_T  GAAC_uncharge     AAC_R  \n",
       "0                   0.701461  0.003851       0.059050 -0.034981  \n",
       "1                   1.465759  0.010432      -0.035113 -0.030800  \n",
       "2                   1.065708  0.018277       0.057837 -0.010125  \n",
       "3                   0.931587 -0.013581      -0.068788  0.016882  \n",
       "4                   1.123222  0.013613       0.063332 -0.031444  \n",
       "...                      ...       ...            ...       ...  \n",
       "2205                1.416965 -0.051409      -0.086435 -0.043283  \n",
       "2206                1.373086  0.014872      -0.030743 -0.005786  \n",
       "2207                0.712799  0.015626       0.022831  0.000327  \n",
       "2208                0.840163 -0.021430       0.020557  0.004958  \n",
       "2209                0.822097 -0.017268       0.003654 -0.062369  \n",
       "\n",
       "[2210 rows x 31 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_features_df = df[[feature for feature in selected_features]]\n",
    "\n",
    "#concatenates original feature vector back into the dataframe\n",
    "df = pd.concat([df.iloc[:, :11], best_features_df], axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "af9f23cf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1878, 38)\n",
      "(332, 38)\n",
      "(1596, 38)\n",
      "(282, 38)\n"
     ]
    }
   ],
   "source": [
    "#choosing 80/20 split instead of 85/15 because of volume of data\n",
    "\n",
    "dev, test = sklearn.model_selection.train_test_split(df, test_size=0.15, random_state=1)\n",
    "\n",
    "train, val = sklearn.model_selection.train_test_split(dev, test_size=0.15, random_state=1)\n",
    "\n",
    "print(dev.shape)\n",
    "print(test.shape)\n",
    "print(train.shape)\n",
    "print(val.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "8470550d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#ID target and features, separate into separate arrays\n",
    "\n",
    "target = 'protein_match'\n",
    "input_features = [columns for columns in df]\n",
    "input_features.remove(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "1ab62bff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1878, 37) (332, 37) (1878, 1) (332, 1)\n"
     ]
    }
   ],
   "source": [
    "#split X and y\n",
    "\n",
    "dev_X = dev[input_features].values\n",
    "test_X = test[input_features].values\n",
    "\n",
    "dev_y = dev[target].values.reshape(-1,1)\n",
    "test_y = test[target].values.reshape(-1,1)  \n",
    "\n",
    "print(dev_X.shape, test_X.shape, dev_y.shape, test_y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "9ac52ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#same thing for training and validation data\n",
    "\n",
    "train_X = train[input_features].values\n",
    "val_X = val[input_features].values\n",
    "\n",
    "train_y = train[target].values.reshape(-1,1)\n",
    "val_y = val[target].values.reshape(-1,1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c374f2d7",
   "metadata": {},
   "source": [
    "Scale the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "2b48b2e5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "scaler = sklearn.preprocessing.StandardScaler()\n",
    "dev_X = scaler.fit_transform(dev_X)\n",
    "test_X = scaler.fit_transform(test_X)\n",
    "train_X = scaler.fit_transform(train_X)\n",
    "val_X = scaler.fit_transform(val_X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3022371",
   "metadata": {},
   "source": [
    "Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "be4e09b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-3 {color: black;background-color: white;}#sk-container-id-3 pre{padding: 0;}#sk-container-id-3 div.sk-toggleable {background-color: white;}#sk-container-id-3 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-3 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-3 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-3 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-3 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-3 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-3 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-3 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-3 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-3 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-3 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-3 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-3 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-3 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-3 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-3 div.sk-item {position: relative;z-index: 1;}#sk-container-id-3 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-3 div.sk-item::before, #sk-container-id-3 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-3 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-3 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-3 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-3 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-3 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-3 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-3 div.sk-label-container {text-align: center;}#sk-container-id-3 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-3 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-3\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-3\" type=\"checkbox\" checked><label for=\"sk-estimator-id-3\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(max_features=0.5, max_samples=0.5, min_samples_split=10,\n",
       "                       min_weight_fraction_leaf=0.000215, n_estimators=150)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Random Forest\n",
    "\n",
    "#hyperparameters determiend with optuna\n",
    "model = sklearn.ensemble.RandomForestClassifier(n_estimators=150, max_depth=None, max_samples=0.5,\n",
    "                                                max_features=0.5, min_weight_fraction_leaf=0.000215,\n",
    "                                               min_samples_split=10)\n",
    "\n",
    "model.fit(train_X, train_y.ravel())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ebf269",
   "metadata": {},
   "source": [
    "Test the model, report relevant statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c73327e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model score is: 0.5780141843971631\n",
      "['No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes'\n",
      " 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No'\n",
      " 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No'\n",
      " 'Yes' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes'\n",
      " 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes'\n",
      " 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'Yes'\n",
      " 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes'\n",
      " 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'No'\n",
      " 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No' 'No' 'Yes' 'Yes'\n",
      " 'Yes' 'No' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No'\n",
      " 'No' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes'\n",
      " 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'No'\n",
      " 'No' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'No'\n",
      " 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes'\n",
      " 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'No'\n",
      " 'Yes' 'No' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes' 'No' 'No' 'No'\n",
      " 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'Yes'\n",
      " 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'No'\n",
      " 'No' 'No' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'No'\n",
      " 'Yes' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'Yes' 'Yes'\n",
      " 'Yes' 'No' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'Yes' 'No' 'No' 'No'\n",
      " 'Yes' 'No' 'Yes' 'No' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'No' 'No' 'Yes'\n",
      " 'Yes' 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes'\n",
      " 'Yes' 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'No' 'No' 'Yes' 'No' 'No' 'Yes'\n",
      " 'No' 'Yes' 'No' 'Yes' 'Yes' 'Yes' 'Yes' 'Yes' 'No' 'No' 'No' 'No' 'Yes'\n",
      " 'Yes' 'Yes' 'Yes' 'No' 'No' 'No']\n"
     ]
    }
   ],
   "source": [
    "score = model.score(val_X, val_y)\n",
    "print('Model score is: {}'.format(score))\n",
    "\n",
    "preds = model.predict(test_X)\n",
    "print(preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "4e9dce62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay at 0x7ff6211ba520>"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#confusion matrix\n",
    "\n",
    "confusion_matrix = sklearn.metrics.confusion_matrix(preds, test_y)\n",
    "sklearn.metrics.ConfusionMatrixDisplay(confusion_matrix).plot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f838ddbc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
